// lint/dist/index.js
#!/usr/bin/env node

// src/index.ts
import { realpathSync } from "fs";
import { pathToFileURL } from "url";
import { confirm as confirm2, input as input2, password } from "@inquirer/prompts";
import chalk12 from "chalk";
import { Command } from "commander";

// src/ai/client.ts
import fs2 from "fs";
import path3 from "path";
import Anthropic from "@anthropic-ai/sdk";
import { input } from "@inquirer/prompts";
import chalk from "chalk";

// src/config.ts
import os from "os";
import path from "path";
var VERSION = true ? "1.2.2" : "0.0.0-dev";
var API_BASE_URL = process.env.LINT_API_URL || "https://api.lint.to";
var HOME_DIR = os.homedir();
var LINT_DIR = ".lint";
var REFS_DIR = path.join(HOME_DIR, LINT_DIR, "refs");
var USERNAME_PATH = path.join(REFS_DIR, "user");
var TOKEN_PATH = path.join(REFS_DIR, "token");
var AI_CONFIG_PATH = path.join(HOME_DIR, LINT_DIR, "ai-config");
var SUPPORTED_EXTENSIONS = {
  eslint: [".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs"],
  prettier: [
    ".js",
    ".jsx",
    ".ts",
    ".tsx",
    ".css",
    ".scss",
    ".less",
    ".json",
    ".html",
    ".vue",
    ".yaml",
    ".yml",
    ".md"
  ],
  rubocop: [".rb", ".rake", ".gemspec"],
  erblint: [".erb"],
  brakeman: [".rb"],
  stylelint: [".css", ".scss", ".sass", ".less"],
  pylint: [".py"],
  biome: [".js", ".jsx", ".ts", ".tsx", ".json", ".css"],
  ruff: [".py", ".pyi"],
  oxlint: [".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs"]
};

// src/utils.ts
import { spawnSync } from "child_process";
import fs from "fs";
import path2 from "path";
import yaml from "js-yaml";
function execFile(command, args, options = {}) {
  const result = spawnSync(command, args, {
    cwd: options.cwd,
    encoding: "utf-8",
    env: options.env ? { ...process.env, ...options.env } : process.env,
    input: options.input,
    stdio: options.silent ? "pipe" : ["pipe", "pipe", "pipe"],
    timeout: 12e4
  });
  if (result.error) {
    const error2 = result.error;
    error2.stdout = result.stdout?.toString() || "";
    error2.stderr = result.stderr?.toString() || "";
    error2.status = result.status;
    throw error2;
  }
  const okExitCodes = new Set(options.okExitCodes ?? [0]);
  if (okExitCodes.has(result.status ?? 0)) {
    return (result.stdout || "").toString().trim();
  }
  const error = new Error(
    (result.stderr || "").toString().trim() || `${command} exited with code ${result.status}`
  );
  error.stdout = (result.stdout || "").toString().trim();
  error.stderr = (result.stderr || "").toString().trim();
  error.status = result.status;
  throw error;
}
function isCommandAvailable(command) {
  try {
    const result = spawnSync("which", [command], { stdio: "pipe" });
    return result.status === 0;
  } catch {
    try {
      const result = spawnSync(command, ["--version"], { stdio: "pipe" });
      return result.status === 0;
    } catch {
      return false;
    }
  }
}
function execGit(args, cwd, options = {}) {
  return execFile("git", args, { ...options, cwd });
}
function findGitDir(startDir) {
  const root = findGitRoot(startDir);
  if (!root) return null;
  try {
    const gitDir = execGit(["rev-parse", "--git-dir"], root, { silent: true });
    return path2.resolve(root, gitDir);
  } catch {
    return null;
  }
}
function findGitRoot(startDir) {
  try {
    return execGit(["rev-parse", "--show-toplevel"], startDir || process.cwd(), { silent: true });
  } catch {
    return null;
  }
}
function repoIsDirty(startDir) {
  const root = findGitRoot(startDir);
  if (!root) return null;
  try {
    const output = execGit(["status", "--short"], root, { silent: true });
    return output.length > 0;
  } catch {
    return null;
  }
}
function getDotLintDir() {
  const gitRoot = findGitRoot();
  if (!gitRoot) return null;
  return path2.join(gitRoot, ".lint");
}
function ensureDir(dirPath) {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }
}
function readLintConfig() {
  const dotLint = getDotLintDir();
  if (!dotLint) return null;
  const configPath = path2.join(dotLint, "config");
  if (!fs.existsSync(configPath)) return null;
  try {
    const content = fs.readFileSync(configPath, "utf-8");
    return yaml.load(content);
  } catch {
    return null;
  }
}
function writeLintConfig(config) {
  const dotLint = getDotLintDir();
  if (!dotLint) throw new Error("Not inside a git repository");
  ensureDir(dotLint);
  const configPath = path2.join(dotLint, "config");
  fs.writeFileSync(configPath, yaml.dump(config), "utf-8");
}
function cleanTmpDir() {
  const dotLint = getDotLintDir();
  if (!dotLint) return;
  const tmpDir = path2.join(dotLint, "tmp");
  if (fs.existsSync(tmpDir)) {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}
function ensureTmpDir() {
  const dotLint = getDotLintDir();
  if (!dotLint) throw new Error("Not inside a git repository");
  const tmpDir = path2.join(dotLint, "tmp");
  ensureDir(tmpDir);
  return tmpDir;
}
function getFileExtension(filePath) {
  return path2.extname(filePath).toLowerCase();
}
function filterFilesByExtensions(files, extensions) {
  return files.filter((file) => extensions.includes(getFileExtension(file)));
}
function formatDuration(ms) {
  if (ms < 1e3) return `${ms}ms`;
  if (ms < 6e4) return `${(ms / 1e3).toFixed(1)}s`;
  return `${Math.floor(ms / 6e4)}m ${Math.round(ms % 6e4 / 1e3)}s`;
}

// src/ai/client.ts
var clientInstance = null;
function getApiKey() {
  const envKey = process.env.ANTHROPIC_API_KEY;
  if (envKey) return envKey;
  try {
    const config = JSON.parse(fs2.readFileSync(AI_CONFIG_PATH, "utf-8"));
    return config.api_key || null;
  } catch {
    return null;
  }
}
function saveApiKey(apiKey) {
  ensureDir(path3.dirname(AI_CONFIG_PATH));
  fs2.writeFileSync(AI_CONFIG_PATH, JSON.stringify({ api_key: apiKey }), "utf-8");
}
function getClient() {
  if (clientInstance) return clientInstance;
  const apiKey = getApiKey();
  if (!apiKey) return null;
  clientInstance = new Anthropic({ apiKey });
  return clientInstance;
}
async function ensureClient() {
  const client = getClient();
  if (client) return client;
  console.log(chalk.yellow("No Anthropic API key found."));
  console.log("Get one at https://console.anthropic.com/\n");
  const apiKey = await input({
    message: "Enter your Anthropic API key:"
  });
  if (!apiKey.startsWith("sk-")) {
    throw new Error("Invalid API key format. Keys start with 'sk-'.");
  }
  saveApiKey(apiKey);
  clientInstance = new Anthropic({ apiKey });
  console.log(chalk.green("API key saved.\n"));
  return clientInstance;
}
async function chat(systemPrompt, userMessage, options) {
  const client = await ensureClient();
  const maxTokens = options?.maxTokens || 4096;
  if (options?.stream) {
    let result = "";
    const stream = await client.messages.stream({
      model: "claude-sonnet-4-20250514",
      max_tokens: maxTokens,
      system: systemPrompt,
      messages: [{ role: "user", content: userMessage }]
    });
    for await (const event of stream) {
      if (event.type === "content_block_delta" && event.delta.type === "text_delta") {
        process.stdout.write(event.delta.text);
        result += event.delta.text;
      }
    }
    console.log();
    return result;
  }
  const response = await client.messages.create({
    model: "claude-sonnet-4-20250514",
    max_tokens: maxTokens,
    system: systemPrompt,
    messages: [{ role: "user", content: userMessage }]
  });
  const textBlock = response.content.find((block) => block.type === "text");
  return textBlock ? textBlock.text : "";
}

// src/ai/commit.ts
import chalk6 from "chalk";

// src/git.ts
import fs6 from "fs";
import path6 from "path";
import { checkbox, confirm } from "@inquirer/prompts";
import chalk5 from "chalk";

// src/api.ts
async function request(endpoint, options = {}) {
  const url = new URL(endpoint, API_BASE_URL);
  if (options.token) {
    url.searchParams.set("user_token", options.token);
  }
  try {
    const response = await fetch(url.toString(), {
      method: options.method || "GET",
      headers: options.body ? { "Content-Type": "application/json" } : void 0,
      body: options.body ? JSON.stringify(options.body) : void 0
    });
    if (!response.ok) {
      const text = await response.text();
      return { error: text || response.statusText, status: response.status };
    }
    const data = await response.json();
    return { data, status: response.status };
  } catch (error) {
    return { error: error.message, status: 0 };
  }
}
async function login(username, password2) {
  return request("/users/sign_in.json", {
    method: "POST",
    body: {
      user: { login: username, password: password2 }
    }
  });
}
async function signup(username, email, password2) {
  return request("/users.json", {
    method: "POST",
    body: {
      user: { username, email, password: password2 }
    }
  });
}
async function fetchUser(username, token) {
  return request(`/${username}.json`, { token });
}
async function searchRepository(username, slug, token) {
  const url = `/${username}/repositories.json`;
  const fullUrl = new URL(url, API_BASE_URL);
  fullUrl.searchParams.set("user_token", token);
  fullUrl.searchParams.set("slug", slug);
  return request(fullUrl.pathname + fullUrl.search);
}
async function createRepository(username, token, name, policy, hasAutofix) {
  return request(`/${username}/repositories.json`, {
    method: "POST",
    token,
    body: {
      repository: { name, policy, has_autofix: hasAutofix }
    }
  });
}
async function fetchPolicy(repositoryUUID, token) {
  return request(`/${repositoryUUID}/policy.json`, { token });
}
async function createCommitAttempt(repositoryUUID, token, sha, branch, message) {
  return request(`/${repositoryUUID}/commit_attempts.json`, {
    method: "POST",
    token,
    body: {
      commit_attempt: { sha, branch, message }
    }
  });
}
async function postReport(token, report) {
  return request("/policy_checks.json", {
    method: "POST",
    token,
    body: { policy_check: report }
  });
}

// src/auth.ts
import fs3 from "fs";
import chalk2 from "chalk";
function getUsername() {
  try {
    return fs3.readFileSync(USERNAME_PATH, "utf-8").trim();
  } catch {
    return null;
  }
}
function getToken() {
  try {
    return fs3.readFileSync(TOKEN_PATH, "utf-8").trim();
  } catch {
    return null;
  }
}
function isLoggedIn() {
  return !!(getUsername() && getToken());
}
function saveCredentials(username, token) {
  ensureDir(REFS_DIR);
  fs3.writeFileSync(USERNAME_PATH, username, "utf-8");
  fs3.writeFileSync(TOKEN_PATH, token, "utf-8");
}
function clearCredentials() {
  try {
    if (fs3.existsSync(USERNAME_PATH)) fs3.unlinkSync(USERNAME_PATH);
    if (fs3.existsSync(TOKEN_PATH)) fs3.unlinkSync(TOKEN_PATH);
  } catch {
  }
}
async function login2(username, password2) {
  const result = await login(username, password2);
  if (result.data) {
    saveCredentials(result.data.username, result.data.authentication_token);
    console.log(`Logged in as ${chalk2.green(result.data.username)}.`);
    return true;
  }
  console.log(chalk2.red("Login failed."), result.error || "Invalid credentials.");
  return false;
}
async function signup2(username, email, password2) {
  const result = await signup(username, email, password2);
  if (result.data) {
    saveCredentials(result.data.username, result.data.authentication_token);
    console.log(`Account created. Logged in as ${chalk2.green(result.data.username)}.`);
    return true;
  }
  console.log(chalk2.red("Signup failed."), result.error || "Unable to create account.");
  return false;
}
function logout() {
  clearCredentials();
  console.log("Logged out.");
}
async function printStatus() {
  const username = getUsername();
  const token = getToken();
  if (!username || !token) {
    console.log(`Not logged in. Run ${chalk2.cyan("lint auth login")} to sign in.`);
    return;
  }
  const result = await fetchUser(username, token);
  if (result.data) {
    console.log(`Logged in as ${chalk2.green(username)}.`);
  } else {
    console.log(`Stored as ${chalk2.yellow(username)} (unable to verify \u2014 API may be offline).`);
  }
}

// src/detect.ts
import fs4 from "fs";
import path4 from "path";
import chalk3 from "chalk";
var SCAN_IGNORED_DIRS = /* @__PURE__ */ new Set([
  ".git",
  ".idea",
  ".next",
  ".nuxt",
  ".venv",
  ".yarn",
  ".pnpm-store",
  "build",
  "coverage",
  "dist",
  "node_modules",
  "tmp",
  "vendor"
]);
var DETECTION_RULES = [
  {
    files: ["package.json", "tsconfig.json", ".nvmrc", "node_modules"],
    language: "JavaScript/TypeScript",
    linters: ["biome"]
  },
  {
    files: ["pyproject.toml", "setup.py", "requirements.txt", "Pipfile", ".python-version"],
    language: "Python",
    linters: ["ruff"]
  },
  {
    files: ["Gemfile", ".ruby-version", "Rakefile", ".rubocop.yml"],
    language: "Ruby",
    linters: ["rubocop"]
  },
  {
    files: [".stylelintrc", ".stylelintrc.json", "postcss.config.js"],
    language: "CSS/SCSS",
    linters: ["stylelint"]
  }
];
var FRAMEWORK_SIGNALS = [
  {
    files: ["config/routes.rb", "bin/rails", "Gemfile"],
    name: "Rails",
    extraLinters: ["brakeman", "erblint"]
  },
  {
    files: ["next.config.js", "next.config.ts", "next.config.mjs"],
    name: "Next.js",
    extraLinters: []
  },
  { files: ["nuxt.config.ts", "nuxt.config.js"], name: "Nuxt", extraLinters: [] },
  { files: ["django", "manage.py"], name: "Django", extraLinters: [] }
];
function detectProject(rootDir) {
  const root = rootDir || findGitRoot() || process.cwd();
  const result = {
    languages: [],
    packageManagers: [],
    frameworks: [],
    hasHusky: false,
    hasLefthook: false
  };
  const discoveredPaths = /* @__PURE__ */ new Set();
  const discoveredExtensions = /* @__PURE__ */ new Set();
  function walk(dir) {
    for (const entry of fs4.readdirSync(dir, { withFileTypes: true })) {
      if (entry.name.startsWith(".") && entry.name !== ".ruby-version" && entry.name !== ".python-version" && entry.name !== ".stylelintrc" && entry.name !== ".stylelintrc.json" && entry.name !== ".rubocop.yml" && entry.name !== ".nvmrc") {
        continue;
      }
      const fullPath = path4.join(dir, entry.name);
      const relativePath = path4.relative(root, fullPath).replaceAll(path4.sep, "/");
      if (entry.isDirectory()) {
        if (!SCAN_IGNORED_DIRS.has(entry.name)) {
          discoveredPaths.add(relativePath);
          walk(fullPath);
        }
        continue;
      }
      if (entry.isFile()) {
        discoveredPaths.add(relativePath);
        discoveredExtensions.add(path4.extname(entry.name).toLowerCase());
      }
    }
  }
  try {
    walk(root);
  } catch {
  }
  for (const rule of DETECTION_RULES) {
    for (const file of rule.files) {
      if (discoveredPaths.has(file) || fs4.existsSync(path4.join(root, file))) {
        if (!result.languages.some((l) => l.name === rule.language)) {
          result.languages.push({
            name: rule.language,
            reason: file,
            suggestedLinters: [...rule.linters]
          });
        }
        break;
      }
    }
  }
  if ((discoveredExtensions.has(".css") || discoveredExtensions.has(".scss") || discoveredExtensions.has(".sass") || discoveredExtensions.has(".less")) && !result.languages.some((l) => l.name === "CSS/SCSS")) {
    result.languages.push({
      name: "CSS/SCSS",
      reason: "stylesheets found",
      suggestedLinters: ["stylelint"]
    });
  }
  if (discoveredExtensions.has(".erb") && !result.languages.some((l) => l.name === "ERB")) {
    result.languages.push({
      name: "ERB",
      reason: "*.erb files found",
      suggestedLinters: ["erblint"]
    });
  }
  for (const fw of FRAMEWORK_SIGNALS) {
    for (const file of fw.files) {
      if (discoveredPaths.has(file) || fs4.existsSync(path4.join(root, file))) {
        if (!result.frameworks.includes(fw.name)) {
          result.frameworks.push(fw.name);
        }
        const preferredLanguage = fw.name === "Rails" ? ["Ruby", "ERB"] : fw.name === "Django" ? ["Python"] : ["JavaScript/TypeScript"];
        for (const linter of fw.extraLinters) {
          const lang = result.languages.find((candidate) => preferredLanguage.includes(candidate.name)) || result.languages.find((candidate) => candidate.suggestedLinters.length > 0);
          if (lang && !lang.suggestedLinters.includes(linter)) {
            lang.suggestedLinters.push(linter);
          }
        }
        break;
      }
    }
  }
  if (fs4.existsSync(path4.join(root, "pnpm-lock.yaml"))) result.packageManagers.push("pnpm");
  else if (fs4.existsSync(path4.join(root, "yarn.lock"))) result.packageManagers.push("yarn");
  else if (fs4.existsSync(path4.join(root, "package-lock.json"))) result.packageManagers.push("npm");
  if (fs4.existsSync(path4.join(root, "Gemfile.lock"))) result.packageManagers.push("bundler");
  if (fs4.existsSync(path4.join(root, "Pipfile.lock"))) result.packageManagers.push("pipenv");
  if (fs4.existsSync(path4.join(root, "poetry.lock"))) result.packageManagers.push("poetry");
  result.hasHusky = fs4.existsSync(path4.join(root, ".husky"));
  result.hasLefthook = fs4.existsSync(path4.join(root, "lefthook.yml")) || fs4.existsSync(path4.join(root, ".lefthook.yml"));
  return result;
}
function getAllSuggestedLinters(project) {
  const linters = /* @__PURE__ */ new Set();
  for (const lang of project.languages) {
    for (const l of lang.suggestedLinters) {
      linters.add(l);
    }
  }
  return [...linters];
}
function checkLinterInstallation(linters) {
  const commands = {
    eslint: "eslint",
    prettier: "prettier",
    rubocop: "rubocop",
    erblint: "erblint",
    brakeman: "brakeman",
    stylelint: "stylelint",
    pylint: "pylint",
    biome: "biome",
    ruff: "ruff",
    oxlint: "oxlint"
  };
  return linters.map((name) => ({
    name,
    installed: isCommandAvailable(commands[name])
  }));
}
function buildSuggestedLinterPlan(project, installStatus = checkLinterInstallation(getAllSuggestedLinters(project))) {
  const reasonsByLinter = /* @__PURE__ */ new Map();
  for (const language of project.languages) {
    for (const linter of language.suggestedLinters) {
      const existing = reasonsByLinter.get(linter) || /* @__PURE__ */ new Set();
      existing.add(`${language.name} (${language.reason})`);
      reasonsByLinter.set(linter, existing);
    }
  }
  return installStatus.map((entry) => ({
    name: entry.name,
    installed: entry.installed,
    reasons: [...reasonsByLinter.get(entry.name) || /* @__PURE__ */ new Set()].sort()
  })).sort((a, b) => a.name.localeCompare(b.name));
}
function printDetectionSummary(project) {
  if (project.languages.length === 0) {
    console.log(chalk3.yellow("  No languages detected."));
    return;
  }
  for (const lang of project.languages) {
    console.log(`  ${chalk3.cyan(lang.name)} ${chalk3.gray(`(${lang.reason})`)}`);
  }
  if (project.frameworks.length > 0) {
    console.log(`  Frameworks: ${project.frameworks.map((f) => chalk3.magenta(f)).join(", ")}`);
  }
}

// src/rc.ts
import fs5 from "fs";
import path5 from "path";
import yaml2 from "js-yaml";
var RC_FILENAMES = [".lintrc.yaml", ".lintrc.yml", ".lintrc.json"];
var LINTER_REPLACEMENTS = {
  biome: ["eslint", "prettier", "oxlint"],
  oxlint: ["eslint"],
  ruff: ["pylint"]
};
function getLinterReplacements() {
  return LINTER_REPLACEMENTS;
}
function findRCFile() {
  const gitRoot = findGitRoot() || process.cwd();
  for (const name of RC_FILENAMES) {
    const filePath = path5.join(gitRoot, name);
    if (fs5.existsSync(filePath)) return filePath;
  }
  return null;
}
function loadRC() {
  const filePath = findRCFile();
  if (!filePath) return {};
  try {
    const content = fs5.readFileSync(filePath, "utf-8");
    if (filePath.endsWith(".json")) {
      return JSON.parse(content);
    }
    return yaml2.load(content) || {};
  } catch {
    return {};
  }
}
function writeRC(rc) {
  const gitRoot = findGitRoot() || process.cwd();
  const filePath = path5.join(gitRoot, ".lintrc.yaml");
  const content = yaml2.dump(rc, { lineWidth: 120, noRefs: true });
  fs5.writeFileSync(filePath, content, "utf-8");
}
function formatRC(rc) {
  return yaml2.dump(rc, { lineWidth: 120, noRefs: true });
}
function resolveEnabledLinters(rc, allLinterNames) {
  let enabled;
  if (rc.linters?.enabled) {
    enabled = rc.linters.enabled;
  } else if (rc.linters?.disabled) {
    enabled = allLinterNames.filter((n) => !rc.linters?.disabled?.includes(n));
  } else {
    enabled = autoResolveConflicts(allLinterNames);
  }
  return enabled;
}
function autoResolveConflicts(available) {
  const disabled = /* @__PURE__ */ new Set();
  for (const [modern, legacy] of Object.entries(LINTER_REPLACEMENTS)) {
    if (available.includes(modern)) {
      for (const old of legacy) {
        if (available.includes(old)) {
          disabled.add(old);
        }
      }
    }
  }
  return available.filter((n) => !disabled.has(n));
}
function shouldIgnoreFile(filePath, patterns) {
  for (const pattern of patterns) {
    const regex = new RegExp(
      `^${pattern.replace(/\./g, "\\.").replace(/\*\*/g, "{{GLOBSTAR}}").replace(/\*/g, "[^/]*").replace(/\{\{GLOBSTAR\}\}/g, ".*")}$`
    );
    if (regex.test(filePath)) return true;
  }
  return false;
}
function filterIgnoredFiles(files, patterns) {
  if (patterns.length === 0) return files;
  return files.filter((f) => !shouldIgnoreFile(f, patterns));
}
function getIgnoredFileDecisions(files, patterns) {
  if (patterns.length === 0) return [];
  const ignored = [];
  for (const file of files) {
    const matchedPattern = patterns.find((pattern) => shouldIgnoreFile(file, [pattern]));
    if (matchedPattern) {
      ignored.push({
        path: file,
        reason: `matched ignore pattern '${matchedPattern}'`
      });
    }
  }
  return ignored;
}
function generateDefaultRC(linters) {
  return {
    linters: {
      enabled: linters
    },
    ignore: ["node_modules/**", "dist/**", "build/**", "coverage/**", ".git/**"],
    fix: {
      enabled: true,
      strategy: "formatter-first"
    },
    hooks: {
      timeout: 60,
      skip_env: "LINT_SKIP"
    }
  };
}
function buildRecommendedRC(existing, linters) {
  const base = generateDefaultRC(linters);
  return {
    linters: {
      enabled: linters,
      ...existing.linters?.disabled ? { disabled: existing.linters.disabled } : {}
    },
    ignore: [.../* @__PURE__ */ new Set([...existing.ignore || [], ...base.ignore || []])],
    fix: {
      enabled: existing.fix?.enabled ?? base.fix?.enabled,
      strategy: existing.fix?.strategy ?? base.fix?.strategy
    },
    hooks: {
      timeout: existing.hooks?.timeout ?? base.hooks?.timeout,
      skip_env: existing.hooks?.skip_env ?? base.hooks?.skip_env
    },
    output: existing.output,
    rules: existing.rules
  };
}

// src/reporter.ts
import chalk4 from "chalk";
import Table from "cli-table3";
var LINT_JSON_SCHEMA_VERSION = "1";
function resolveJsonOutcome(reports, meta) {
  const totalErrors = reports.reduce((s, r) => s + r.error_count, 0);
  const totalWarnings = reports.reduce((s, r) => s + r.warning_count, 0);
  if (reports.length === 0 && meta.message) {
    return { status: "skipped", exitCode: 0 };
  }
  if (totalErrors > 0) {
    return { status: "failed", exitCode: 1 };
  }
  if (totalWarnings > 0 && meta.failOnWarnings) {
    return { status: "failed_on_warnings", exitCode: 2 };
  }
  if (totalWarnings > 0) {
    return { status: "passed_with_warnings", exitCode: 0 };
  }
  return { status: "passed", exitCode: 0 };
}
function printReport(reports, truncate = false) {
  for (const report of reports) {
    if (report.error_count === 0 && report.warning_count === 0) continue;
    console.log(chalk4.bold(`
\u2500\u2500\u2500 ${report.linter.toUpperCase()} \u2500\u2500\u2500
`));
    for (const file of report.files) {
      if (file.offenses.length === 0) continue;
      console.log(chalk4.underline(file.path));
      const offenses = truncate ? file.offenses.slice(0, 10) : file.offenses;
      for (const offense of offenses) {
        const severity = offense.severity === "error" ? chalk4.red("error") : offense.severity === "warning" ? chalk4.yellow("warn ") : chalk4.blue("info ");
        const location = chalk4.gray(`${offense.line}:${offense.column}`);
        const rule = chalk4.gray(offense.rule);
        const fixable = offense.fixable ? chalk4.green(" [fixable]") : "";
        console.log(`  ${location}  ${severity}  ${offense.message}  ${rule}${fixable}`);
      }
      if (truncate && file.offenses.length > 10) {
        console.log(chalk4.gray(`  ... and ${file.offenses.length - 10} more`));
      }
      console.log("");
    }
  }
}
function printSummaryTable(reports) {
  const table = new Table({
    head: [
      chalk4.white("Linter"),
      chalk4.red("Errors"),
      chalk4.yellow("Warnings"),
      chalk4.green("Fixable")
    ],
    style: { head: [], border: [] }
  });
  for (const report of reports) {
    if (report.error_count === 0 && report.warning_count === 0) continue;
    table.push([
      report.linter,
      report.error_count.toString(),
      report.warning_count.toString(),
      (report.fixable_error_count + report.fixable_warning_count).toString()
    ]);
  }
  if (table.length > 0) {
    console.log(table.toString());
  }
}
function formatJsonReport(reports, meta) {
  const totalErrors = reports.reduce((s, r) => s + r.error_count, 0);
  const totalWarnings = reports.reduce((s, r) => s + r.warning_count, 0);
  const totalFixable = reports.reduce(
    (s, r) => s + r.fixable_error_count + r.fixable_warning_count,
    0
  );
  const outcome = resolveJsonOutcome(reports, meta);
  return JSON.stringify(
    {
      schema_version: LINT_JSON_SCHEMA_VERSION,
      kind: "lint_run",
      success: outcome.exitCode === 0,
      status: outcome.status,
      exit_code: outcome.exitCode,
      summary: {
        errors: totalErrors,
        warnings: totalWarnings,
        fixable: totalFixable,
        duration_ms: meta.duration,
        dry_run: meta.dryRun ?? false,
        auto_fix: meta.fix ?? false
      },
      run: {
        cwd: meta.cwd ?? process.cwd(),
        mode: meta.mode ?? "staged files",
        file_count: meta.fileCount ?? reports.reduce((sum, report) => sum + report.files.length, 0),
        linters: meta.linterNames ?? reports.map((report) => report.linter),
        policy_rule_count: meta.policyRuleCount ?? 0,
        requested_paths: meta.requestedPaths ?? []
      },
      decisions: {
        ignored_files: meta.ignoredFiles ?? [],
        linter_selection: meta.linterSelection ?? [],
        file_coverage: meta.fileCoverage ?? {
          coveredFiles: [],
          uncoveredFiles: []
        },
        policy: meta.policySummary ?? {
          source: "local",
          totalRules: meta.policyRuleCount ?? 0,
          applicableRules: meta.policyRuleCount ?? 0,
          byLinter: {}
        }
      },
      ...meta.message ? { message: meta.message } : {},
      linters: reports.map((r) => ({
        name: r.linter,
        errors: r.error_count,
        warnings: r.warning_count,
        fixable: r.fixable_error_count + r.fixable_warning_count,
        files: r.files.map((f) => ({
          path: f.path,
          offenses: f.offenses.map((o) => ({
            line: o.line,
            column: o.column,
            rule: o.rule,
            message: o.message,
            severity: o.severity,
            fixable: o.fixable ?? false
          }))
        }))
      }))
    },
    null,
    2
  );
}
function formatRunDecisionReport(report) {
  const lines = [];
  lines.push("  Lint Explain Run");
  lines.push("");
  lines.push(`  CWD: ${report.cwd}`);
  lines.push(`  Mode: ${report.mode}`);
  lines.push(
    `  Files: ${report.lintableFileCount} lintable / ${report.discoveredFileCount} discovered`
  );
  lines.push(
    `  Policy: ${report.policy.source} (${report.policy.applicableRules}/${report.policy.totalRules} applicable rules)`
  );
  lines.push(
    `  Fix strategy: ${report.fixStrategy.autofix ? "autofix enabled" : "check only"} (${report.fixStrategy.strategy})`
  );
  if (report.conflicts.length > 0) {
    lines.push("");
    lines.push("  Resolved Conflicts:");
    lines.push("");
    for (const conflict of report.conflicts) {
      lines.push(
        `    \u2713 ${conflict.winner} over ${conflict.losers.join(", ")} (${conflict.reason})`
      );
    }
  }
  lines.push("");
  lines.push("  Selected Linters:");
  lines.push("");
  for (const linter of report.linterSelection.filter((entry) => entry.selected)) {
    lines.push(`    \u2713 ${linter.name} (${linter.reason})`);
  }
  const skippedLinters = report.linterSelection.filter((entry) => !entry.selected);
  if (skippedLinters.length > 0) {
    lines.push("");
    lines.push("  Skipped Linters:");
    lines.push("");
    for (const linter of skippedLinters) {
      lines.push(`    - ${linter.name} (${linter.reason})`);
    }
  }
  if (report.ignoredFiles.length > 0) {
    lines.push("");
    lines.push("  Ignored Files:");
    lines.push("");
    for (const file of report.ignoredFiles.slice(0, 10)) {
      lines.push(`    - ${file.path} (${file.reason})`);
    }
    if (report.ignoredFiles.length > 10) {
      lines.push(`    ... and ${report.ignoredFiles.length - 10} more`);
    }
  }
  if (report.fileCoverage.coveredFiles.length > 0) {
    lines.push("");
    lines.push("  Covered Files:");
    lines.push("");
    for (const file of report.fileCoverage.coveredFiles.slice(0, 10)) {
      lines.push(`    \u2713 ${file.path} -> ${file.linters.join(", ")} (${file.reason})`);
    }
    if (report.fileCoverage.coveredFiles.length > 10) {
      lines.push(`    ... and ${report.fileCoverage.coveredFiles.length - 10} more`);
    }
  }
  if (report.fileCoverage.uncoveredFiles.length > 0) {
    lines.push("");
    lines.push("  Uncovered Files:");
    lines.push("");
    for (const file of report.fileCoverage.uncoveredFiles.slice(0, 10)) {
      lines.push(`    - ${file.path} (${file.reason})`);
    }
    if (report.fileCoverage.uncoveredFiles.length > 10) {
      lines.push(`    ... and ${report.fileCoverage.uncoveredFiles.length - 10} more`);
    }
  }
  if (Object.keys(report.policy.byLinter).length > 0) {
    lines.push("");
    lines.push("  Policy Rules By Linter:");
    lines.push("");
    for (const [linter, count] of Object.entries(report.policy.byLinter)) {
      lines.push(`    ${linter}: ${count}`);
    }
  }
  if (report.nextSteps.length > 0) {
    lines.push("");
    lines.push("  Recommended Next Steps:");
    lines.push("");
    for (const step of report.nextSteps) {
      lines.push(`    - ${step}`);
    }
  }
  lines.push("");
  return lines;
}
function formatRunDecisionJson(report) {
  return JSON.stringify(
    {
      schema_version: LINT_JSON_SCHEMA_VERSION,
      kind: "lint_explain_run",
      ...report
    },
    null,
    2
  );
}
function formatMachineSummaryJson(report) {
  return JSON.stringify(
    {
      schema_version: LINT_JSON_SCHEMA_VERSION,
      kind: "lint_machine_summary",
      ...report
    },
    null,
    2
  );
}

// src/git.ts
var MANAGED_HOOK_MARKER = "Managed by Lint";
var INSTALL_COMMANDS = {
  biome: "npm install -g @biomejs/biome",
  eslint: "npm install -g eslint",
  prettier: "npm install -g prettier",
  oxlint: "npm install -g oxlint",
  stylelint: "npm install -g stylelint",
  ruff: "pip install ruff",
  pylint: "pip install pylint",
  rubocop: "gem install rubocop",
  erblint: "gem install erb_lint",
  brakeman: "gem install brakeman"
};
function buildBootstrapPlan(args) {
  const rc = generateDefaultRC(args.suggestedLinters);
  return {
    repoName: args.repoName,
    selectedLinters: args.suggestedLinters,
    missingLinters: args.installStatus.filter((entry) => !entry.installed && args.suggestedLinters.includes(entry.name)).map((entry) => entry.name),
    rcExists: args.rcExists,
    lintConfigExists: args.lintConfigExists,
    hookTimeout: rc.hooks?.timeout ?? 60,
    hookSkipEnv: rc.hooks?.skip_env ?? "LINT_SKIP"
  };
}
function buildSetupFixPlan(args) {
  const bootstrapPlan = buildBootstrapPlan(args);
  return {
    ...bootstrapPlan,
    recommendedLinters: args.suggestedLinters,
    willWriteRecommendedConfig: true,
    willCreateLintConfig: !args.lintConfigExists,
    willInstallMissing: args.installMissing && bootstrapPlan.missingLinters.length > 0,
    willInstallHooks: args.installHooks
  };
}
function getStagedFiles() {
  try {
    const output = execGit(
      ["diff", "--cached", "--name-status", "-z", "--diff-filter=AMDR"],
      void 0,
      {
        silent: true
      }
    );
    if (!output) return [];
    const entries = output.split("\0").filter(Boolean);
    const files = [];
    for (let i = 0; i < entries.length; ) {
      const statusCode = entries[i++] || "";
      let status;
      let filePath;
      if (statusCode.startsWith("A")) {
        status = "added";
        filePath = entries[i++] || "";
      } else if (statusCode.startsWith("D")) {
        status = "deleted";
        filePath = entries[i++] || "";
      } else if (statusCode.startsWith("R")) {
        status = "renamed";
        i += 1;
        filePath = entries[i++] || "";
      } else {
        status = "modified";
        filePath = entries[i++] || "";
      }
      if (filePath) {
        files.push({ path: filePath, status });
      }
    }
    return files;
  } catch {
    return [];
  }
}
function getStagedFilePaths(excludeDeleted = true) {
  const files = getStagedFiles();
  if (excludeDeleted) {
    return files.filter((f) => f.status !== "deleted").map((f) => f.path);
  }
  return files.map((f) => f.path);
}
function getCurrentSha() {
  try {
    return execGit(["rev-parse", "HEAD"], void 0, { silent: true });
  } catch {
    return "unknown";
  }
}
function getCurrentBranch() {
  try {
    return execGit(["rev-parse", "--abbrev-ref", "HEAD"], void 0, { silent: true });
  } catch {
    return "unknown";
  }
}
function getStagedDiff() {
  try {
    return execGit(["diff", "--cached"], void 0, { silent: true });
  } catch {
    return "";
  }
}
function findFiles(targetPath) {
  const resolved = path6.resolve(targetPath);
  if (!fs6.existsSync(resolved)) {
    return [];
  }
  if (fs6.statSync(resolved).isFile()) {
    return [resolved];
  }
  const files = [];
  const ignoreDirs = /* @__PURE__ */ new Set([
    "node_modules",
    ".git",
    "dist",
    "build",
    "coverage",
    ".lint",
    "__pycache__"
  ]);
  function walk(dir) {
    for (const entry of fs6.readdirSync(dir, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        if (!ignoreDirs.has(entry.name) && !entry.name.startsWith(".")) {
          walk(path6.join(dir, entry.name));
        }
      } else if (entry.isFile()) {
        files.push(path6.join(dir, entry.name));
      }
    }
  }
  walk(resolved);
  return files;
}
function hookContent(lintInvocation, npxInvocation, timeout, skipEnv) {
  return `#!/bin/sh
# ${MANAGED_HOOK_MARKER}. Reinstall through:
#   lint hooks install
# Skip: ${skipEnv}=1 git commit ...
# Skip: git commit --no-verify

set -eu

[ "$${skipEnv}" = "1" ] && exit 0

run_with_timeout() {
  TIMEOUT_SECONDS="$1"
  shift
  TIMEOUT_FLAG="\${TMPDIR:-/tmp}/lint-hook-timeout.$$"
  rm -f "$TIMEOUT_FLAG"

  "$@" &
  CMD_PID=$!

  (
    sleep "$TIMEOUT_SECONDS"
    if kill -0 "$CMD_PID" 2>/dev/null; then
      : > "$TIMEOUT_FLAG"
      kill "$CMD_PID" 2>/dev/null || true
    fi
  ) &
  WATCHER_PID=$!

  wait "$CMD_PID" || STATUS=$?
  STATUS=\${STATUS:-0}
  kill "$WATCHER_PID" 2>/dev/null || true
  wait "$WATCHER_PID" 2>/dev/null || true

  if [ -f "$TIMEOUT_FLAG" ]; then
    rm -f "$TIMEOUT_FLAG"
    echo "Lint: hook timed out after ${timeout}s." >&2
    return 124
  fi

  rm -f "$TIMEOUT_FLAG"
  return "$STATUS"
}

if command -v lint >/dev/null 2>&1; then
  ${lintInvocation}
elif command -v npx >/dev/null 2>&1; then
  ${npxInvocation}
else
  echo "Lint: lint command not found. Skipping hook."
  exit 0
fi
`;
}
function inspectHook(hooksDir, name) {
  const hookPath = path6.join(hooksDir, name);
  if (!fs6.existsSync(hookPath)) {
    return { name, hookPath, exists: false, managed: false };
  }
  const content = fs6.readFileSync(hookPath, "utf-8");
  return {
    name,
    hookPath,
    exists: true,
    managed: content.includes(MANAGED_HOOK_MARKER)
  };
}
function inspectManagedHooks(gitRoot) {
  const gitDir = findGitDir(gitRoot);
  if (!gitDir) return [];
  return ["pre-commit", "prepare-commit-msg", "post-commit"].map(
    (name) => inspectHook(path6.join(gitDir, "hooks"), name)
  );
}
function installHooks(options) {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk5.red("Not inside a git repository."));
    return;
  }
  const timeout = options?.timeout ?? 60;
  const skipEnv = options?.skipEnv ?? "LINT_SKIP";
  const gitDir = findGitDir(gitRoot);
  if (!gitDir) {
    console.log(chalk5.red("Unable to locate .git directory."));
    return;
  }
  const hasHusky = fs6.existsSync(path6.join(gitRoot, ".husky"));
  const hasLefthook = fs6.existsSync(path6.join(gitRoot, "lefthook.yml")) || fs6.existsSync(path6.join(gitRoot, ".lefthook.yml"));
  if (hasHusky) {
    console.log(chalk5.yellow("  Husky detected. Adding Lint as a Husky hook."));
    const huskyDir = path6.join(gitRoot, ".husky");
    const hookPath = path6.join(huskyDir, "pre-commit");
    const existing = fs6.existsSync(hookPath) ? fs6.readFileSync(hookPath, "utf-8") : "";
    if (!existing.includes("lint")) {
      const content = `${existing.trimEnd()}
lint pre-commit
`;
      fs6.writeFileSync(hookPath, content, { mode: 493 });
      console.log(chalk5.green("  \u2713 Added to .husky/pre-commit"));
    } else {
      console.log(chalk5.gray("  Already in .husky/pre-commit"));
    }
    return;
  }
  if (hasLefthook) {
    console.log(chalk5.yellow("  Lefthook detected. Add Lint manually to lefthook.yml:"));
    console.log(
      chalk5.gray("    pre-commit:\n      commands:\n        lint:\n          run: lint pre-commit")
    );
    return;
  }
  const hooksDir = path6.join(gitDir, "hooks");
  ensureDir(hooksDir);
  const hooks = {
    "pre-commit": hookContent(
      `run_with_timeout "${timeout}" lint pre-commit -t "$@"`,
      `run_with_timeout "${timeout}" npx --no-install lint pre-commit -t "$@"`,
      timeout,
      skipEnv
    ),
    "prepare-commit-msg": hookContent(
      `run_with_timeout "${timeout}" lint prepare-commit-msg "$@"`,
      `run_with_timeout "${timeout}" npx --no-install lint prepare-commit-msg "$@"`,
      timeout,
      skipEnv
    ),
    "post-commit": hookContent(
      `run_with_timeout "${timeout}" lint post-commit "$@"`,
      `run_with_timeout "${timeout}" npx --no-install lint post-commit "$@"`,
      timeout,
      skipEnv
    )
  };
  for (const [name, content] of Object.entries(hooks)) {
    const hookPath = path6.join(hooksDir, name);
    const inspection = inspectHook(hooksDir, name);
    if (inspection.exists && !inspection.managed) {
      const backupDir = path6.join(hooksDir, `backup_${Date.now()}`);
      ensureDir(backupDir);
      fs6.copyFileSync(hookPath, path6.join(backupDir, name));
    }
    fs6.writeFileSync(hookPath, content, { mode: 493 });
    console.log(chalk5.green(`  \u2713 ${name}`));
  }
  console.log(chalk5.green("\nGit hooks installed."));
  console.log(chalk5.gray(`  Timeout: ${timeout}s | Skip: ${skipEnv}=1 git commit ...`));
}
function uninstallHooks() {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk5.red("Not inside a git repository."));
    return;
  }
  const gitDir = findGitDir(gitRoot);
  if (!gitDir) {
    console.log(chalk5.red("Unable to locate .git directory."));
    return;
  }
  const hooksDir = path6.join(gitDir, "hooks");
  const hookNames = ["pre-commit", "prepare-commit-msg", "post-commit"];
  for (const name of hookNames) {
    const inspection = inspectHook(hooksDir, name);
    if (inspection.exists && inspection.managed) {
      fs6.unlinkSync(inspection.hookPath);
      console.log(chalk5.yellow(`  \u2717 ${name} removed`));
    }
  }
  console.log("Git hooks uninstalled.");
}
async function init() {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk5.red("Not inside a git repository. Run 'git init' first."));
    return;
  }
  const repoName = path6.basename(gitRoot);
  console.log(chalk5.cyan.bold("\n  Lint Setup\n"));
  const existing = readLintConfig();
  if (existing?.uuid) {
    console.log(chalk5.yellow(`Already initialized (${existing.uuid}).`));
    const reinit = await confirm({ message: "Re-initialize?" });
    if (!reinit) return;
  }
  console.log(chalk5.bold("Scanning project...\n"));
  const project = detectProject(gitRoot);
  printDetectionSummary(project);
  console.log("");
  const suggested = getAllSuggestedLinters(project);
  const installStatus = checkLinterInstallation(suggested);
  const linterChoices = installStatus.map((l) => ({
    name: `${l.name} ${l.installed ? chalk5.green("(installed)") : chalk5.red("(not installed)")}`,
    value: l.name,
    checked: true
  }));
  let selectedLinters;
  if (linterChoices.length > 0) {
    selectedLinters = await checkbox({
      message: "Which linters should Lint use?",
      choices: linterChoices
    });
  } else {
    console.log(chalk5.yellow("No linters detected for this project."));
    selectedLinters = [];
  }
  const missing = installStatus.filter((l) => !l.installed && selectedLinters.includes(l.name));
  if (missing.length > 0) {
    const installMissing = await confirm({
      message: `Install ${missing.length} missing linter(s)? (${missing.map((l) => l.name).join(", ")})`,
      default: true
    });
    if (installMissing) {
      for (const l of missing) {
        const cmd = INSTALL_COMMANDS[l.name];
        if (cmd) {
          try {
            process.stdout.write(chalk5.gray(`  Installing ${l.name}... `));
            execFile("sh", ["-lc", cmd], { silent: true });
            console.log(chalk5.green("done"));
          } catch {
            console.log(chalk5.red("failed"));
          }
        }
      }
      console.log("");
    }
  }
  const rc = generateDefaultRC(selectedLinters);
  writeRC(rc);
  console.log(chalk5.green("  \u2713 Created .lintrc.yaml"));
  const installHooksAnswer = await confirm({
    message: "Install git hooks? (auto-lint on commit)",
    default: true
  });
  if (installHooksAnswer) {
    installHooks({
      timeout: rc.hooks?.timeout ?? 60,
      skipEnv: rc.hooks?.skip_env ?? "LINT_SKIP"
    });
  }
  const config = { uuid: `local-${Date.now()}`, repository: repoName };
  if (isLoggedIn()) {
    const username = getUsername();
    const token = getToken();
    const searchResult = await searchRepository(username, repoName, token);
    if (searchResult.data?.uuid) {
      config.uuid = searchResult.data.uuid;
      config.username = username;
      console.log(chalk5.green(`  \u2713 Connected to cloud: ${repoName}`));
    } else {
      const connectCloud = await confirm({
        message: "Create repository on Lint cloud? (team policies)",
        default: false
      });
      if (connectCloud) {
        const createResult = await createRepository(
          username,
          token,
          repoName,
          "recommended",
          true
        );
        if (createResult.data?.uuid) {
          config.uuid = createResult.data.uuid;
          config.username = username;
          console.log(chalk5.green(`  \u2713 Created on cloud: ${repoName}`));
        }
      }
    }
  }
  writeLintConfig(config);
  const gitignorePath = path6.join(gitRoot, ".gitignore");
  if (fs6.existsSync(gitignorePath)) {
    const gitignore = fs6.readFileSync(gitignorePath, "utf-8");
    if (!gitignore.includes(".lint/tmp")) {
      fs6.appendFileSync(gitignorePath, "\n# Lint\n.lint/tmp/\n", "utf-8");
    }
  }
  console.log(chalk5.green.bold("\n  \u2713 Lint initialized!\n"));
  console.log(`  Run ${chalk5.cyan("lint")} to lint staged files.`);
  console.log(`  Run ${chalk5.cyan("lint .")} to lint the entire project.`);
  console.log(`  Run ${chalk5.cyan("lint ai review")} for AI code review.
`);
}
function bootstrapProject(options) {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk5.red("Not inside a git repository. Run 'git init' first."));
    return;
  }
  const repoName = path6.basename(gitRoot);
  const project = detectProject(gitRoot);
  const suggested = getAllSuggestedLinters(project);
  const installStatus = checkLinterInstallation(suggested);
  const plan = buildBootstrapPlan({
    repoName,
    suggestedLinters: suggested,
    installStatus,
    rcExists: fs6.existsSync(path6.join(gitRoot, ".lintrc.yaml")),
    lintConfigExists: Boolean(readLintConfig()?.uuid)
  });
  if (options?.json) {
    console.log(
      JSON.stringify(
        {
          schema_version: LINT_JSON_SCHEMA_VERSION,
          kind: "lint_bootstrap",
          ...plan,
          install_missing_requested: options.installMissing ?? false,
          install_hooks_requested: options.installHooks ?? false,
          dry_run: options.dryRun ?? false
        },
        null,
        2
      )
    );
    return;
  }
  console.log(chalk5.cyan.bold("\n  Lint Bootstrap\n"));
  printDetectionSummary(project);
  console.log("");
  console.log(`  Repo: ${repoName}`);
  console.log(`  Suggested linters: ${plan.selectedLinters.join(", ") || "none"}`);
  console.log(`  Missing linters: ${plan.missingLinters.join(", ") || "none"}`);
  console.log(`  RC config: ${plan.rcExists ? "update" : "create"}`);
  console.log(`  Lint config: ${plan.lintConfigExists ? "keep existing" : "create local config"}`);
  console.log("");
  if (options?.dryRun) return;
  writeRC(generateDefaultRC(plan.selectedLinters));
  console.log(chalk5.green("  \u2713 Wrote .lintrc.yaml"));
  if (!plan.lintConfigExists) {
    writeLintConfig({ uuid: `local-${Date.now()}`, repository: repoName });
    console.log(chalk5.green("  \u2713 Wrote .lint/config"));
  }
  if (options?.installMissing && plan.missingLinters.length > 0) {
    for (const name of plan.missingLinters) {
      const cmd = INSTALL_COMMANDS[name];
      if (!cmd) continue;
      try {
        process.stdout.write(chalk5.gray(`  Installing ${name}... `));
        execFile("sh", ["-lc", cmd], { silent: true });
        console.log(chalk5.green("done"));
      } catch {
        console.log(chalk5.red("failed"));
      }
    }
  }
  if (options?.installHooks) {
    installHooks({ timeout: plan.hookTimeout, skipEnv: plan.hookSkipEnv });
  }
  console.log(chalk5.green("\nBootstrap complete."));
}
function fixSetup(options) {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk5.red("Not inside a git repository. Run 'git init' first."));
    return;
  }
  const repoName = path6.basename(gitRoot);
  const project = detectProject(gitRoot);
  const suggestedPlan = buildSuggestedLinterPlan(project);
  const suggestedLinters = suggestedPlan.map((entry) => entry.name);
  const existingRc = loadRC();
  const recommendedRc = buildRecommendedRC(existingRc, suggestedLinters);
  const plan = buildSetupFixPlan({
    repoName,
    suggestedLinters,
    installStatus: suggestedPlan.map((entry) => ({ name: entry.name, installed: entry.installed })),
    rcExists: fs6.existsSync(path6.join(gitRoot, ".lintrc.yaml")),
    lintConfigExists: Boolean(readLintConfig()?.uuid),
    installMissing: options?.installMissing ?? true,
    installHooks: options?.installHooks ?? true
  });
  if (options?.json) {
    console.log(
      JSON.stringify(
        {
          schema_version: LINT_JSON_SCHEMA_VERSION,
          kind: "lint_setup_fix",
          ...plan,
          dry_run: options.dryRun ?? false,
          recommended_config: recommendedRc,
          suggested_linter_reasons: suggestedPlan
        },
        null,
        2
      )
    );
    return;
  }
  console.log(chalk5.cyan.bold("\n  Lint Setup Fix\n"));
  printDetectionSummary(project);
  console.log("");
  console.log(`  Repo: ${plan.repoName}`);
  console.log(`  Recommended linters: ${plan.recommendedLinters.join(", ") || "none"}`);
  console.log(`  Write recommended config: ${plan.willWriteRecommendedConfig ? "yes" : "no"}`);
  console.log(`  Create local lint config: ${plan.willCreateLintConfig ? "yes" : "no"}`);
  console.log(`  Install missing linters: ${plan.willInstallMissing ? "yes" : "no"}`);
  console.log(`  Install hooks: ${plan.willInstallHooks ? "yes" : "no"}`);
  console.log("");
  if (options?.dryRun) return;
  writeRC(recommendedRc);
  console.log(chalk5.green("  \u2713 Wrote recommended .lintrc.yaml"));
  if (plan.willCreateLintConfig) {
    writeLintConfig({ uuid: `local-${Date.now()}`, repository: repoName });
    console.log(chalk5.green("  \u2713 Wrote .lint/config"));
  }
  if (plan.willInstallMissing) {
    for (const name of plan.missingLinters) {
      const cmd = INSTALL_COMMANDS[name];
      if (!cmd) continue;
      try {
        process.stdout.write(chalk5.gray(`  Installing ${name}... `));
        execFile("sh", ["-lc", cmd], { silent: true });
        console.log(chalk5.green("done"));
      } catch {
        console.log(chalk5.red("failed"));
      }
    }
  }
  if (plan.willInstallHooks) {
    installHooks({ timeout: plan.hookTimeout, skipEnv: plan.hookSkipEnv });
  }
  console.log(chalk5.green("\nSetup fix complete."));
}

// src/ai/commit.ts
var SYSTEM_PROMPT = `You are a git commit message generator. Given a diff of staged changes, write a concise, conventional commit message.

Rules:
- Use conventional commit format: type(scope): description
- Types: feat, fix, refactor, docs, style, test, chore, perf, ci, build
- Keep the first line under 72 characters
- Add a blank line then a brief body (2-3 bullet points) if the change is complex
- Be specific: "fix null check in user auth" not "fix bug"
- Never start with "Update" or "Change" \u2014 describe what was actually done

Output ONLY the commit message, nothing else.`;
async function generateCommitMessage() {
  if (!getApiKey()) return null;
  const files = getStagedFilePaths();
  if (files.length === 0) return null;
  const diff = getStagedDiff();
  if (!diff) return null;
  const truncatedDiff = diff.length > 8e3 ? `${diff.substring(0, 8e3)}
... (truncated)` : diff;
  const userMessage = `Files: ${files.join(", ")}

Diff:
\`\`\`diff
${truncatedDiff}
\`\`\``;
  try {
    const message = await chat(SYSTEM_PROMPT, userMessage, { maxTokens: 256 });
    return message.trim();
  } catch {
    return null;
  }
}
async function printCommitSuggestion() {
  console.log(chalk6.cyan("Generating commit message from staged changes...\n"));
  const message = await generateCommitMessage();
  if (!message) {
    console.log(
      chalk6.yellow("Could not generate commit message. Check your API key with 'lint ai setup'.")
    );
    return;
  }
  console.log(chalk6.bold("Suggested commit message:\n"));
  console.log(`  ${chalk6.green(message.split("\n").join("\n  "))}`);
  console.log(chalk6.gray(`
  Copy: git commit -m "${message.split("\n")[0]}"`));
}

// src/ai/explain.ts
import chalk7 from "chalk";
var SYSTEM_PROMPT2 = `You are a friendly code mentor working as part of Lint, a universal linting tool.
Your job is to explain linting errors in plain language so developers can learn from them.

For each error/warning:
1. Explain what the rule means and why it exists
2. Show what's wrong in the code
3. Show how to fix it with a code example
4. Rate importance: \u{1F534} Must fix, \u{1F7E1} Should fix, \u{1F535} Nice to fix

Be concise but educational. Use simple language.`;
async function explainErrors(reports) {
  const allOffenses = reports.flatMap(
    (r) => r.files.flatMap(
      (f) => f.offenses.map((o) => ({
        linter: r.linter,
        file: f.path,
        ...o
      }))
    )
  );
  if (allOffenses.length === 0) {
    console.log(chalk7.green("No errors to explain \u2014 your code is clean!"));
    return;
  }
  const uniqueRules = [...new Set(allOffenses.map((o) => o.rule))].slice(0, 20);
  const selectedOffenses = uniqueRules.map((rule) => allOffenses.find((o) => o.rule === rule)).filter((o) => o !== void 0);
  console.log(chalk7.cyan(`
Explaining ${selectedOffenses.length} linting issue(s)...
`));
  const errorSummary = selectedOffenses.map((o) => `- [${o.linter}] ${o.rule}: "${o.message}" in ${o.file}:${o.line}`).join("\n");
  const userMessage = `Explain these linting errors:

${errorSummary}`;
  try {
    await chat(SYSTEM_PROMPT2, userMessage, { stream: true, maxTokens: 4096 });
  } catch (error) {
    console.log(chalk7.red("\nAI explain failed:"), error.message);
  }
}

// src/ai/fix.ts
import fs7 from "fs";
import chalk8 from "chalk";
var SYSTEM_PROMPT3 = `You are an expert code fixer working as part of Lint, a universal linting tool.
Given a git diff of staged changes, identify issues and provide fixed code.

For each fix:
1. Explain what the issue is (one line)
2. Show the corrected code block

Only suggest fixes for genuine issues (bugs, security, performance).
Do NOT suggest style changes or subjective improvements.

Format each fix as:
### Fix: <brief description>
**File:** \`path/to/file\`
**Issue:** <explanation>
\`\`\`<language>
<corrected code>
\`\`\``;
async function fixStagedChanges() {
  const files = getStagedFilePaths();
  if (files.length === 0) {
    console.log(chalk8.yellow("No staged files to fix."));
    return;
  }
  const diff = getStagedDiff();
  if (!diff) {
    console.log(chalk8.yellow("No changes detected in staged files."));
    return;
  }
  console.log(
    chalk8.cyan(`
Analyzing ${files.length} staged file(s) for auto-fix suggestions...
`)
  );
  const fileContents = [];
  for (const file of files.slice(0, 10)) {
    try {
      const content = fs7.readFileSync(file, "utf-8");
      if (content.length < 1e4) {
        fileContents.push(`--- ${file} ---
${content}`);
      }
    } catch {
    }
  }
  const userMessage = `Analyze and suggest fixes for these staged changes:

Diff:
\`\`\`diff
${diff}
\`\`\`

${fileContents.length > 0 ? `
Full file contents:
${fileContents.join("\n\n")}` : ""}`;
  try {
    await chat(SYSTEM_PROMPT3, userMessage, { stream: true, maxTokens: 4096 });
  } catch (error) {
    console.log(chalk8.red("\nAI fix failed:"), error.message);
  }
}

// src/ai/review.ts
import chalk9 from "chalk";
var SYSTEM_PROMPT4 = `You are an expert code reviewer working as part of Lint, a universal linting tool.
Your job is to review code changes (git diffs) and provide actionable feedback.

Focus on:
- Bugs and logic errors
- Security vulnerabilities
- Performance issues
- Code quality and maintainability
- Best practices for the language being used

Be concise. Use this format for each issue:
**[severity]** \`file:line\` - description
- Suggestion: how to fix

Where severity is one of: \u{1F534} ERROR, \u{1F7E1} WARNING, \u{1F535} INFO

End with a brief summary.`;
async function reviewStagedChanges() {
  const files = getStagedFilePaths();
  if (files.length === 0) {
    console.log(chalk9.yellow("No staged files to review."));
    return;
  }
  const diff = getStagedDiff();
  if (!diff) {
    console.log(chalk9.yellow("No changes detected in staged files."));
    return;
  }
  console.log(chalk9.cyan(`
Reviewing ${files.length} staged file(s) with AI...
`));
  const userMessage = `Review these staged changes:

Files: ${files.join(", ")}

Diff:
\`\`\`diff
${diff}
\`\`\``;
  try {
    await chat(SYSTEM_PROMPT4, userMessage, { stream: true, maxTokens: 4096 });
  } catch (error) {
    console.log(chalk9.red("\nAI review failed:"), error.message);
  }
}

// src/doctor.ts
var ALL_LINTERS = [
  "biome",
  "oxlint",
  "eslint",
  "prettier",
  "ruff",
  "pylint",
  "rubocop",
  "erblint",
  "brakeman",
  "stylelint"
];
function collectDoctorReport() {
  const gitRoot = findGitRoot();
  const branch = gitRoot ? getCurrentBranch() : null;
  const dirty = gitRoot ? repoIsDirty(gitRoot) : null;
  const project = detectProject(gitRoot || process.cwd());
  const config = readLintConfig();
  const rcFile = findRCFile();
  const rc = loadRC();
  const installedLinters = checkLinterInstallation(ALL_LINTERS);
  const installedNames = installedLinters.filter((linter) => linter.installed).map((linter) => linter.name);
  const projectApplicable = getAllSuggestedLinters(project);
  const projectApplicableSet = new Set(projectApplicable);
  const enabledNames = rc.linters?.enabled ? rc.linters.enabled : rc.linters?.disabled ? projectApplicable.filter((name) => !rc.linters?.disabled?.includes(name)) : autoResolveConflicts(projectApplicable);
  const selectedNames = resolveEnabledLinters(
    rc,
    installedNames.filter((name) => projectApplicableSet.has(name))
  );
  const enabledSet = new Set(enabledNames);
  const disabledSet = new Set(rc.linters?.disabled || []);
  const selectedSet = new Set(selectedNames);
  const source = rc.linters?.enabled || rc.linters?.disabled ? "rc" : "auto";
  const linters = installedLinters.map((linter) => ({
    name: linter.name,
    installed: linter.installed,
    enabled: enabledSet.has(linter.name) && !disabledSet.has(linter.name),
    selected: selectedSet.has(linter.name),
    source
  }));
  const hooks = gitRoot ? inspectManagedHooks(gitRoot).map((hook) => ({
    name: hook.name,
    exists: hook.exists,
    managed: hook.managed,
    hookPath: hook.hookPath
  })) : [];
  const missingSelectedLinters = linters.filter((linter) => linter.enabled && !linter.installed).map((linter) => linter.name);
  const managedHooks = hooks.filter((hook) => hook.exists && hook.managed).length;
  const status = gitRoot && (config?.uuid || rcFile) && missingSelectedLinters.length === 0 ? "healthy" : "needs_setup";
  return {
    schema_version: LINT_JSON_SCHEMA_VERSION,
    kind: "lint_doctor",
    status,
    git: {
      root: gitRoot,
      branch,
      dirty
    },
    config: {
      initialized: Boolean(config?.uuid),
      mode: config?.uuid ? config.uuid.startsWith("local-") ? "offline" : "cloud" : null,
      rcFile,
      loggedIn: isLoggedIn(),
      username: getUsername(),
      hasLintConfig: Boolean(config?.uuid)
    },
    project: {
      languages: project.languages,
      frameworks: project.frameworks,
      packageManagers: project.packageManagers,
      suggestedLinters: getAllSuggestedLinters(project),
      hasHusky: project.hasHusky,
      hasLefthook: project.hasLefthook
    },
    linters,
    hooks,
    summary: {
      installedLinters: linters.filter((linter) => linter.installed).length,
      enabledLinters: linters.filter((linter) => linter.enabled).length,
      selectedLinters: linters.filter((linter) => linter.selected).length,
      missingSelectedLinters,
      managedHooks
    }
  };
}
function formatDoctorReport(report) {
  const lines = [];
  lines.push("  Lint Doctor");
  lines.push("");
  lines.push(`  Status: ${report.status}`);
  lines.push("");
  if (report.git.root) {
    lines.push(`  \u2713 Git: ${report.git.root}`);
    const dirtyLabel = report.git.dirty === null ? "unknown" : report.git.dirty ? "dirty" : "clean";
    lines.push(`  \u2713 Branch: ${report.git.branch || "unknown"} (${dirtyLabel})`);
  } else {
    lines.push("  \u2717 Not a git repository");
  }
  if (report.config.initialized) {
    lines.push(`  \u2713 Config: .lint/config (${report.config.mode})`);
  } else {
    lines.push("  \u2717 No .lint/config \u2014 run 'lint init'");
  }
  lines.push(
    report.config.rcFile ? `  \u2713 RC: ${report.config.rcFile}` : "  - No .lintrc.yaml (using defaults)"
  );
  lines.push(
    report.config.loggedIn && report.config.username ? `  \u2713 Auth: ${report.config.username}` : "  - Not logged in (offline mode)"
  );
  lines.push("");
  lines.push("  Project:");
  lines.push("");
  if (report.project.languages.length === 0) {
    lines.push("    - No languages detected");
  } else {
    for (const language of report.project.languages) {
      const linters = language.suggestedLinters.length > 0 ? ` -> ${language.suggestedLinters.join(", ")}` : "";
      lines.push(`    \u2713 ${language.name} (${language.reason})${linters}`);
    }
  }
  if (report.project.frameworks.length > 0) {
    lines.push(`    Frameworks: ${report.project.frameworks.join(", ")}`);
  }
  if (report.project.packageManagers.length > 0) {
    lines.push(`    Package managers: ${report.project.packageManagers.join(", ")}`);
  }
  if (report.project.hasHusky || report.project.hasLefthook) {
    lines.push(
      `    Hook managers: ${[
        report.project.hasHusky ? "husky" : null,
        report.project.hasLefthook ? "lefthook" : null
      ].filter(Boolean).join(", ")}`
    );
  }
  lines.push("");
  lines.push("  Linters:");
  lines.push("");
  for (const linter of report.linters) {
    const enabled = linter.enabled ? "" : linter.source === "rc" ? " (disabled in .lintrc.yaml)" : " (auto-disabled)";
    const selected = linter.selected ? " [selected]" : "";
    lines.push(
      `    ${linter.installed ? "\u2713" : "\u2717"} ${linter.name}: ${linter.installed ? "installed" : "not installed"}${enabled}${selected}`
    );
  }
  lines.push("");
  lines.push("  Git Hooks:");
  lines.push("");
  for (const hook of report.hooks) {
    if (!hook.exists) {
      lines.push(`    - ${hook.name} (not installed)`);
    } else if (hook.managed) {
      lines.push(`    \u2713 ${hook.name}`);
    } else {
      lines.push(`    ~ ${hook.name} (not Lint)`);
    }
  }
  if (report.summary.missingSelectedLinters.length > 0) {
    lines.push("");
    lines.push(`  Missing selected linters: ${report.summary.missingSelectedLinters.join(", ")}`);
  }
  lines.push("");
  return lines;
}

// src/orchestrator.ts
import chalk11 from "chalk";
import { createSpinner } from "nanospinner";

// src/linters/biome.ts
import fs8 from "fs";
import path7 from "path";

// src/linters/base.ts
import chalk10 from "chalk";
var BaseLinter = class {
  get extensions() {
    return SUPPORTED_EXTENSIONS[this.name] || [];
  }
  isInstalled() {
    return isCommandAvailable(this.command);
  }
  install() {
    console.log(chalk10.cyan(`Installing ${this.name}...`));
    try {
      execFile("sh", ["-lc", this.installCmd]);
      console.log(chalk10.green(`${this.name} installed.`));
    } catch (error) {
      console.log(chalk10.red(`Failed to install ${this.name}.`), error.message);
    }
  }
  selectFiles(stagedFiles) {
    return filterFilesByExtensions(stagedFiles, this.extensions);
  }
  execute(stagedFiles, rules, autofix = false) {
    const files = this.selectFiles(stagedFiles);
    if (files.length === 0) return null;
    if (!this.isInstalled()) {
      console.log(chalk10.yellow(`${this.name} is not installed. Skipping.`));
      return null;
    }
    const tmpDir = ensureTmpDir();
    const configPath = this.createConfig(rules, tmpDir);
    try {
      return this.run(files, configPath, autofix);
    } catch (error) {
      console.log(chalk10.red(`${this.name} failed:`), error.message);
      return {
        success: false,
        report: this.emptyReport()
      };
    }
  }
  emptyReport() {
    return {
      linter: this.name,
      files: [],
      error_count: 0,
      warning_count: 0,
      fixable_error_count: 0,
      fixable_warning_count: 0
    };
  }
  buildReport(files) {
    let errorCount = 0;
    let warningCount = 0;
    let fixableErrorCount = 0;
    let fixableWarningCount = 0;
    for (const file of files) {
      for (const offense of file.offenses) {
        if (offense.severity === "error") {
          errorCount++;
          if (offense.fixable) fixableErrorCount++;
        } else {
          warningCount++;
          if (offense.fixable) fixableWarningCount++;
        }
      }
    }
    return {
      linter: this.name,
      files,
      error_count: errorCount,
      warning_count: warningCount,
      fixable_error_count: fixableErrorCount,
      fixable_warning_count: fixableWarningCount
    };
  }
};

// src/linters/biome.ts
function offsetToLineColumn(source, offset) {
  if (!source || offset <= 0) {
    return { line: 1, column: 1 };
  }
  const clampedOffset = Math.min(offset, source.length);
  let line = 1;
  let column = 1;
  for (let i = 0; i < clampedOffset; i++) {
    if (source[i] === "\n") {
      line++;
      column = 1;
    } else {
      column++;
    }
  }
  return { line, column };
}
var BiomeLinter = class extends BaseLinter {
  name = "biome";
  command = "biome";
  installCmd = "npm install -g @biomejs/biome";
  configFileName = "biome.json";
  createConfig(rules, tmpDir) {
    const biomeRules = rules.filter((r) => r.linter === "biome");
    if (biomeRules.length === 0) return "";
    const config = {
      $schema: "https://biomejs.dev/schemas/1.9.0/schema.json",
      linter: {
        enabled: true,
        rules: { recommended: true }
      },
      formatter: {
        enabled: true,
        indentStyle: "space",
        indentWidth: 2
      }
    };
    const linterRules = {};
    for (const rule of biomeRules) {
      const [group, name] = rule.slug.includes("/") ? rule.slug.split("/") : ["recommended", rule.slug];
      if (!linterRules[group]) linterRules[group] = {};
      linterRules[group][name] = rule.status === "enabled" ? rule.severity : "off";
    }
    if (Object.keys(linterRules).length > 0) {
      config.linter.rules = linterRules;
    }
    const configPath = path7.join(tmpDir, this.configFileName);
    fs8.writeFileSync(configPath, JSON.stringify(config, null, 2), "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    let raw;
    try {
      raw = execFile(
        "biome",
        [
          ...autofix ? ["check", "--write"] : ["lint"],
          ...configPath ? [`--config-path=${path7.dirname(configPath)}`] : [],
          "--reporter=json",
          ...files
        ],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || '{"diagnostics":[]}';
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let output;
    try {
      output = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileMap = /* @__PURE__ */ new Map();
    for (const diag of output.diagnostics || []) {
      const file = diag.location?.path?.file || "unknown";
      const offenses = fileMap.get(file) || [];
      const position = offsetToLineColumn(
        diag.location?.sourceCode,
        diag.location?.span?.start || 0
      );
      offenses.push({
        rule: diag.category || "biome",
        message: diag.message?.content || diag.description || "Unknown issue",
        severity: diag.severity === "error" ? "error" : "warning",
        line: position.line,
        column: position.column,
        source: diag.location?.sourceCode
      });
      fileMap.set(file, offenses);
    }
    const fileReports = Array.from(fileMap.entries()).map(([filePath, offenses]) => ({
      path: filePath,
      offenses
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/brakeman.ts
var BrakemanLinter = class extends BaseLinter {
  name = "brakeman";
  command = "brakeman";
  installCmd = "gem install brakeman";
  configFileName = "";
  // Brakeman doesn't use a config file in the same way
  createConfig(_rules, _tmpDir) {
    return "";
  }
  run(files, _configPath, _autofix) {
    const onlyFiles = files.join(",");
    let raw;
    try {
      raw = execFile("brakeman", ["-f", "json", "--no-pager", "--only-files", onlyFiles], {
        silent: true
      });
    } catch (error) {
      const err = error;
      if (err.status === 4) {
        return { success: true, report: this.emptyReport() };
      }
      raw = err.stdout || '{"warnings":[],"errors":[]}';
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let output;
    try {
      output = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileMap = /* @__PURE__ */ new Map();
    for (const warning of output.warnings) {
      const offenses = fileMap.get(warning.file) || [];
      offenses.push({
        rule: warning.check_name || warning.warning_type,
        message: `[${warning.warning_type}] ${warning.message} (confidence: ${warning.confidence})`,
        severity: warning.confidence === "High" ? "error" : "warning",
        line: warning.line || 1,
        column: 1
      });
      fileMap.set(warning.file, offenses);
    }
    for (const error of output.errors || []) {
      const file = error.location || "unknown";
      const offenses = fileMap.get(file) || [];
      offenses.push({
        rule: "brakeman/error",
        message: error.error,
        severity: "error",
        line: 1,
        column: 1
      });
      fileMap.set(file, offenses);
    }
    const fileReports = Array.from(fileMap.entries()).map(([filePath, offenses]) => ({
      path: filePath,
      offenses
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/erblint.ts
import fs9 from "fs";
import path8 from "path";
import yaml3 from "js-yaml";
var ErbLintLinter = class extends BaseLinter {
  name = "erblint";
  command = "erblint";
  installCmd = "gem install erb_lint";
  configFileName = ".erb-lint.yml";
  createConfig(rules, tmpDir) {
    const erblintRules = rules.filter((r) => r.linter === "erblint");
    if (erblintRules.length === 0) return "";
    const config = {
      linters: {
        Rubocop: { enabled: true }
      }
    };
    for (const rule of erblintRules) {
      if (!config.linters) config.linters = {};
      config.linters[rule.slug] = {
        enabled: rule.status === "enabled"
      };
    }
    const configPath = path8.join(tmpDir, this.configFileName);
    fs9.writeFileSync(configPath, yaml3.dump(config), "utf-8");
    return configPath;
  }
  run(files, configPath, _autofix) {
    let raw;
    try {
      raw = execFile(
        "erblint",
        [...configPath ? ["--config", configPath] : [], "--format", "json", ...files],
        {
          silent: true,
          env: {
            LANG: "en_US.UTF-8",
            LC_ALL: "en_US.UTF-8"
          }
        }
      );
    } catch (error) {
      raw = error.stdout || '{"files":[],"summary":{"offenses":0}}';
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, files) {
    try {
      const output = JSON.parse(raw);
      if (output.files) {
        const fileReports = output.files.filter((f) => f.offenses?.length > 0).map(
          (f) => ({
            path: f.path,
            offenses: f.offenses.map(
              (o) => ({
                rule: o.linter || "erblint",
                message: o.message,
                severity: "warning",
                line: o.location?.start_line || 1,
                column: o.location?.start_column || 1
              })
            )
          })
        );
        return this.buildReport(fileReports);
      }
    } catch {
    }
    const lines = raw.split("\n").filter((l) => l.trim());
    const offenses = [];
    for (const line of lines) {
      const match = line.match(/^(.+):(\d+):(\d+):\s*(.+)$/);
      if (match) {
        offenses.push({
          rule: "erblint",
          message: match[4],
          severity: "warning",
          line: Number.parseInt(match[2], 10),
          column: Number.parseInt(match[3], 10)
        });
      }
    }
    if (offenses.length === 0) return this.emptyReport();
    return this.buildReport([{ path: files[0] || "unknown", offenses }]);
  }
};

// src/linters/eslint.ts
import fs10 from "fs";
import path9 from "path";
var ESLintLinter = class extends BaseLinter {
  name = "eslint";
  command = "eslint";
  installCmd = "npm install -g eslint";
  configFileName = "eslintrc.json";
  createConfig(rules, tmpDir) {
    const eslintPolicyRules = rules.filter((r) => r.linter === "eslint");
    if (eslintPolicyRules.length === 0) return "";
    const config = {
      env: { es2022: true, node: true, browser: true },
      parserOptions: {
        ecmaVersion: "latest",
        sourceType: "module",
        ecmaFeatures: { jsx: true }
      },
      rules: {}
    };
    const eslintRules = {};
    for (const rule of eslintPolicyRules) {
      if (rule.status === "enabled") {
        eslintRules[rule.slug] = rule.severity === "error" ? "error" : "warn";
      } else {
        eslintRules[rule.slug] = "off";
      }
    }
    config.rules = eslintRules;
    const configPath = path9.join(tmpDir, this.configFileName);
    fs10.writeFileSync(configPath, JSON.stringify(config, null, 2), "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    let raw;
    try {
      raw = execFile(
        "eslint",
        [
          ...configPath ? ["--config", configPath] : [],
          "--format",
          "json",
          ...autofix ? ["--fix"] : [],
          ...files
        ],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || "[]";
    }
    const report = this.parseOutput(raw, files);
    return {
      success: report.error_count === 0,
      report,
      raw
    };
  }
  parseOutput(raw, _files) {
    let results;
    try {
      results = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileReports = results.filter((r) => r.messages.length > 0).map((result) => ({
      path: result.filePath,
      offenses: result.messages.map(
        (msg) => ({
          rule: msg.ruleId || "unknown",
          message: msg.message,
          severity: msg.severity === 2 ? "error" : "warning",
          line: msg.line,
          column: msg.column,
          fixable: !!msg.fix
        })
      )
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/oxlint.ts
import fs11 from "fs";
import path10 from "path";
function offsetToLineColumn2(filePath, offset) {
  try {
    const source = fs11.readFileSync(filePath, "utf-8");
    const clampedOffset = Math.min(Math.max(offset, 0), source.length);
    let line = 1;
    let column = 1;
    for (let i = 0; i < clampedOffset; i++) {
      if (source[i] === "\n") {
        line++;
        column = 1;
      } else {
        column++;
      }
    }
    return { line, column };
  } catch {
    return { line: 1, column: 1 };
  }
}
var OxlintLinter = class extends BaseLinter {
  name = "oxlint";
  command = "oxlint";
  installCmd = "npm install -g oxlint";
  configFileName = ".oxlintrc.json";
  createConfig(rules, tmpDir) {
    const oxlintRulesList = rules.filter((r) => r.linter === "oxlint");
    if (oxlintRulesList.length === 0) return "";
    const config = {
      rules: {}
    };
    const oxlintRules = {};
    for (const rule of oxlintRulesList) {
      oxlintRules[rule.slug] = rule.status === "enabled" ? rule.severity : "off";
    }
    config.rules = oxlintRules;
    const configPath = path10.join(tmpDir, this.configFileName);
    fs11.writeFileSync(configPath, JSON.stringify(config, null, 2), "utf-8");
    return configPath;
  }
  run(files, configPath, _autofix) {
    let raw;
    try {
      raw = execFile(
        "oxlint",
        [...configPath ? [`--config=${configPath}`] : [], "--format=json", ...files],
        {
          silent: true
        }
      );
    } catch (error) {
      raw = error.stdout || "[]";
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let diagnostics;
    try {
      diagnostics = JSON.parse(raw);
      if (!Array.isArray(diagnostics)) diagnostics = [];
    } catch {
      return this.emptyReport();
    }
    const fileMap = /* @__PURE__ */ new Map();
    for (const diag of diagnostics) {
      const file = diag.filename || "unknown";
      const offenses = fileMap.get(file) || [];
      const offset = diag.labels[0]?.span?.offset || 0;
      const position = offsetToLineColumn2(file, offset);
      offenses.push({
        rule: diag.code?.code || "oxlint",
        message: diag.message,
        severity: diag.severity === "error" ? "error" : "warning",
        line: position.line,
        column: position.column
      });
      fileMap.set(file, offenses);
    }
    const fileReports = Array.from(fileMap.entries()).map(([filePath, offenses]) => ({
      path: filePath,
      offenses
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/prettier.ts
import fs12 from "fs";
import path11 from "path";
var PrettierLinter = class extends BaseLinter {
  name = "prettier";
  command = "prettier";
  installCmd = "npm install -g prettier";
  configFileName = ".prettierrc.json";
  createConfig(rules, tmpDir) {
    const prettierRules = rules.filter((r) => r.linter === "prettier");
    if (prettierRules.length === 0) return "";
    const config = {};
    for (const rule of prettierRules) {
      if (rule.status === "enabled" && rule.content?.options) {
        Object.assign(config, rule.content.options);
      }
    }
    const configPath = path11.join(tmpDir, this.configFileName);
    fs12.writeFileSync(configPath, JSON.stringify(config, null, 2), "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    if (autofix) {
      try {
        execFile(
          "prettier",
          [...configPath ? ["--config", configPath] : [], "--write", ...files],
          { silent: true }
        );
      } catch {
      }
      return { success: true, report: this.emptyReport() };
    }
    let raw;
    try {
      raw = execFile(
        "prettier",
        [...configPath ? ["--config", configPath] : [], "--check", ...files],
        { silent: true }
      );
      return { success: true, report: this.emptyReport(), raw };
    } catch (error) {
      raw = error.stdout || "";
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    const lines = raw.split("\n").filter((l) => l.trim());
    const unformatted = lines.filter((l) => l.includes("[warn]") || !l.startsWith("Checking"));
    const fileReports = unformatted.filter((line) => !line.startsWith("[warn]")).map((filePath) => ({
      path: filePath.trim(),
      offenses: [
        {
          rule: "prettier/prettier",
          message: "File is not formatted",
          severity: "warning",
          line: 1,
          column: 1,
          fixable: true
        }
      ]
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/pylint.ts
import fs13 from "fs";
import path12 from "path";
var PylintLinter = class extends BaseLinter {
  name = "pylint";
  command = "pylint";
  installCmd = "pip install pylint";
  configFileName = ".pylintrc";
  createConfig(rules, tmpDir) {
    const pylintRules = rules.filter((r) => r.linter === "pylint");
    if (pylintRules.length === 0) return "";
    const sections = {
      MAIN: {},
      BASIC: {},
      FORMAT: { "max-line-length": "120" },
      SIMILARITIES: { "min-similarity-lines": "6" },
      MESSAGES_CONTROL: {}
    };
    const disabled = [];
    const enabled = [];
    for (const rule of pylintRules) {
      if (rule.status === "enabled") {
        enabled.push(rule.slug);
      } else {
        disabled.push(rule.slug);
      }
    }
    if (disabled.length) sections.MESSAGES_CONTROL.disable = disabled.join(",");
    if (enabled.length) sections.MESSAGES_CONTROL.enable = enabled.join(",");
    let content = "";
    for (const [section, values] of Object.entries(sections)) {
      content += `[${section}]
`;
      for (const [key, value] of Object.entries(values)) {
        content += `${key}=${value}
`;
      }
      content += "\n";
    }
    const configPath = path12.join(tmpDir, this.configFileName);
    fs13.writeFileSync(configPath, content, "utf-8");
    return configPath;
  }
  run(files, configPath, _autofix) {
    let raw;
    try {
      raw = execFile(
        "pylint",
        [...configPath ? [`--rcfile=${configPath}`] : [], "--output-format=json", ...files],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || "[]";
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let messages;
    try {
      messages = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileMap = /* @__PURE__ */ new Map();
    for (const msg of messages) {
      const offenses = fileMap.get(msg.path) || [];
      offenses.push({
        rule: `${msg.symbol} (${msg["message-id"]})`,
        message: msg.message,
        severity: msg.type === "error" || msg.type === "fatal" ? "error" : "warning",
        line: msg.line,
        column: msg.column
      });
      fileMap.set(msg.path, offenses);
    }
    const fileReports = Array.from(fileMap.entries()).map(([filePath, offenses]) => ({
      path: filePath,
      offenses
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/rubocop.ts
import fs14 from "fs";
import path13 from "path";
import yaml4 from "js-yaml";
var RuboCopLinter = class extends BaseLinter {
  name = "rubocop";
  command = "rubocop";
  installCmd = "gem install rubocop";
  configFileName = ".rubocop.yml";
  createConfig(rules, tmpDir) {
    const rubocopRules = rules.filter((r) => r.linter === "rubocop");
    if (rubocopRules.length === 0) return "";
    const config = {
      AllCops: { NewCops: "enable", SuggestExtensions: false }
    };
    for (const rule of rubocopRules) {
      config[rule.slug] = {
        Enabled: rule.status === "enabled",
        Severity: rule.severity === "error" ? "error" : "warning"
      };
    }
    const configPath = path13.join(tmpDir, this.configFileName);
    fs14.writeFileSync(configPath, yaml4.dump(config), "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    let raw;
    try {
      raw = execFile(
        "rubocop",
        [
          ...configPath ? ["--config", configPath] : [],
          "--format",
          "json",
          ...autofix ? ["--autocorrect"] : [],
          ...files
        ],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || '{"files":[],"summary":{"offense_count":0}}';
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let output;
    try {
      output = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileReports = output.files.filter((f) => f.offenses.length > 0).map((file) => ({
      path: file.path,
      offenses: file.offenses.map(
        (o) => ({
          rule: o.cop_name,
          message: o.message,
          severity: o.severity === "error" || o.severity === "fatal" ? "error" : "warning",
          line: o.location.start_line,
          column: o.location.start_column,
          fixable: o.correctable
        })
      )
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/ruff.ts
import fs15 from "fs";
import path14 from "path";
var RuffLinter = class extends BaseLinter {
  name = "ruff";
  command = "ruff";
  installCmd = "pip install ruff";
  configFileName = "ruff.toml";
  createConfig(rules, tmpDir) {
    const ruffRules = rules.filter((r) => r.linter === "ruff");
    if (ruffRules.length === 0) return "";
    const selectRules = [];
    const ignoreRules = [];
    for (const rule of ruffRules) {
      if (rule.status === "enabled") {
        selectRules.push(rule.slug);
      } else {
        ignoreRules.push(rule.slug);
      }
    }
    let content = `# Generated by Lint
line-length = 120
target-version = "py312"

[lint]
`;
    if (selectRules.length > 0) {
      content += `select = [${selectRules.map((r) => `"${r}"`).join(", ")}]
`;
    }
    if (ignoreRules.length > 0) {
      content += `ignore = [${ignoreRules.map((r) => `"${r}"`).join(", ")}]
`;
    }
    const configPath = path14.join(tmpDir, this.configFileName);
    fs15.writeFileSync(configPath, content, "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    let raw;
    try {
      raw = execFile(
        "ruff",
        [
          ...autofix ? ["check", "--fix"] : ["check"],
          ...configPath ? [`--config=${configPath}`] : [],
          "--output-format=json",
          ...files
        ],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || "[]";
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let diagnostics;
    try {
      diagnostics = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileMap = /* @__PURE__ */ new Map();
    for (const diag of diagnostics) {
      const offenses = fileMap.get(diag.filename) || [];
      offenses.push({
        rule: diag.code,
        message: diag.message,
        severity: diag.code.startsWith("E") || diag.code.startsWith("F") ? "error" : "warning",
        line: diag.location.row,
        column: diag.location.column,
        fixable: !!diag.fix
      });
      fileMap.set(diag.filename, offenses);
    }
    const fileReports = Array.from(fileMap.entries()).map(([filePath, offenses]) => ({
      path: filePath,
      offenses
    }));
    return this.buildReport(fileReports);
  }
};

// src/linters/stylelint.ts
import fs16 from "fs";
import path15 from "path";
var StylelintLinter = class extends BaseLinter {
  name = "stylelint";
  command = "stylelint";
  installCmd = "npm install -g stylelint";
  configFileName = ".stylelintrc.json";
  createConfig(rules, tmpDir) {
    const stylelintPolicyRules = rules.filter((r) => r.linter === "stylelint");
    if (stylelintPolicyRules.length === 0) return "";
    const config = { rules: {} };
    const stylelintRules = {};
    for (const rule of stylelintPolicyRules) {
      if (rule.status === "enabled") {
        stylelintRules[rule.slug] = [true, { severity: rule.severity }];
      } else {
        stylelintRules[rule.slug] = null;
      }
    }
    config.rules = stylelintRules;
    const configPath = path15.join(tmpDir, this.configFileName);
    fs16.writeFileSync(configPath, JSON.stringify(config, null, 2), "utf-8");
    return configPath;
  }
  run(files, configPath, autofix) {
    let raw;
    try {
      raw = execFile(
        "stylelint",
        [
          ...configPath ? ["--config", configPath] : [],
          "-f",
          "json",
          ...autofix ? ["--fix"] : [],
          ...files
        ],
        { silent: true }
      );
    } catch (error) {
      raw = error.stdout || "[]";
    }
    const report = this.parseOutput(raw, files);
    return { success: report.error_count === 0, report, raw };
  }
  parseOutput(raw, _files) {
    let results;
    try {
      results = JSON.parse(raw);
    } catch {
      return this.emptyReport();
    }
    const fileReports = results.filter((r) => r.warnings.length > 0).map((result) => ({
      path: result.source,
      offenses: result.warnings.map(
        (w) => ({
          rule: w.rule,
          message: w.text,
          severity: w.severity === "error" ? "error" : "warning",
          line: w.line,
          column: w.column
        })
      )
    }));
    return this.buildReport(fileReports);
  }
};

// src/orchestrator.ts
var ALL_LINTERS2 = [
  new BiomeLinter(),
  new OxlintLinter(),
  new ESLintLinter(),
  new PrettierLinter(),
  new RuboCopLinter(),
  new ErbLintLinter(),
  new BrakemanLinter(),
  new StylelintLinter(),
  new RuffLinter(),
  new PylintLinter()
];
var LINTER_MAP = new Map(
  ALL_LINTERS2.map((l) => [l.name, l])
);
function projectApplicableLinterSet() {
  const project = detectProject(process.cwd());
  return new Set(getAllSuggestedLinters(project));
}
function selectLinters(rc) {
  const installed = ALL_LINTERS2.filter((l) => l.isInstalled());
  const applicable = projectApplicableLinterSet();
  const candidateNames = installed.map((l) => l.name).filter((name) => rc.linters?.enabled || applicable.has(name));
  const enabledNames = resolveEnabledLinters(rc, candidateNames);
  return installed.filter((l) => enabledNames.includes(l.name));
}
function describeLinterSelection(rc) {
  const installedNames = ALL_LINTERS2.filter((linter) => linter.isInstalled()).map(
    (linter) => linter.name
  );
  const applicable = projectApplicableLinterSet();
  const candidateNames = installedNames.filter(
    (name) => rc.linters?.enabled || applicable.has(name)
  );
  const selectedNames = new Set(resolveEnabledLinters(rc, candidateNames));
  return ALL_LINTERS2.map((linter) => {
    const name = linter.name;
    const installed = installedNames.includes(name);
    const enabled = rc.linters?.enabled ? rc.linters.enabled.includes(name) : rc.linters?.disabled ? !rc.linters.disabled.includes(name) : installed;
    let reason = "available";
    if (!installed) {
      reason = "not installed";
    } else if (!selectedNames.has(name)) {
      reason = rc.linters?.enabled || rc.linters?.disabled ? "filtered by .lintrc" : "auto-resolved conflict";
    } else if (rc.linters?.enabled || rc.linters?.disabled) {
      reason = "selected by .lintrc";
    }
    return {
      name,
      installed,
      enabled,
      selected: selectedNames.has(name),
      reason
    };
  });
}
function describeFileCoverage(files, selection, linters) {
  const selectionByName = new Map(selection.map((entry) => [entry.name, entry]));
  const coverage = files.map((file) => {
    const selectedLinters = linters.filter((linter) => linter.selectFiles([file]).length > 0).map((linter) => linter.name).sort();
    const supportedByKnownLinters = ALL_LINTERS2.filter(
      (linter) => linter.selectFiles([file]).length > 0
    ).map((linter) => linter.name).sort();
    const supportedInstalledLinters = supportedByKnownLinters.filter(
      (name) => selectionByName.get(name)?.installed
    );
    const supportedEnabledLinters = supportedByKnownLinters.filter(
      (name) => selectionByName.get(name)?.enabled
    );
    let reason = "selected linter matched this file";
    if (selectedLinters.length === 0) {
      if (supportedByKnownLinters.length === 0) {
        reason = "no known linter supports this file type";
      } else if (supportedInstalledLinters.length === 0) {
        reason = `supported by ${supportedByKnownLinters.join(", ")}, but those linters are not installed`;
      } else if (supportedEnabledLinters.length === 0) {
        reason = `supported by ${supportedInstalledLinters.join(", ")}, but disabled by configuration`;
      } else {
        reason = `supported by ${supportedEnabledLinters.join(", ")}, but none were selected`;
      }
    } else if (selectedLinters.length > 1) {
      reason = "multiple selected linters matched this file";
    }
    return {
      path: file,
      linters: selectedLinters,
      reason
    };
  });
  return {
    coveredFiles: coverage.filter((entry) => entry.linters.length > 0),
    uncoveredFiles: coverage.filter((entry) => entry.linters.length === 0).map((entry) => ({
      path: entry.path,
      reason: entry.reason
    }))
  };
}
function summarizePolicyRules(policyRules, linters) {
  const selectedLinterNames = new Set(linters.map((linter) => linter.name));
  const byLinter = Object.fromEntries(
    Object.entries(
      policyRules.reduce((counts, rule) => {
        counts[rule.linter] = (counts[rule.linter] || 0) + 1;
        return counts;
      }, {})
    ).sort(([a], [b]) => a.localeCompare(b))
  );
  return {
    source: policyRules.length > 0 ? "cloud" : "local",
    totalRules: policyRules.length,
    applicableRules: policyRules.filter((rule) => selectedLinterNames.has(rule.linter)).length,
    byLinter
  };
}
function describeResolvedConflicts(selection) {
  const replacements = getLinterReplacements();
  const selectionByName = new Map(selection.map((entry) => [entry.name, entry]));
  const conflicts = [];
  for (const [winner, losers] of Object.entries(replacements)) {
    const winnerSelection = selectionByName.get(winner);
    if (!winnerSelection?.selected) continue;
    const resolvedLosers = losers.filter((loser) => {
      const loserSelection = selectionByName.get(loser);
      return loserSelection?.installed && !loserSelection.selected;
    });
    if (resolvedLosers.length > 0) {
      conflicts.push({
        winner,
        losers: resolvedLosers,
        reason: "modern linter replacement rule"
      });
    }
  }
  return conflicts;
}
function describeNextSteps(report) {
  const steps = [];
  const missingLinters = report.linterSelection.filter((entry) => entry.reason === "not installed");
  if (missingLinters.length > 0) {
    steps.push(
      `Install missing linters: ${missingLinters.map((entry) => entry.name).join(", ")} (or run 'lint install missing .')`
    );
  }
  const uncoveredByMissingInstall = report.fileCoverage.uncoveredFiles.some(
    (file) => file.reason.includes("not installed")
  );
  if (uncoveredByMissingInstall) {
    steps.push("Re-run after installing the missing linters to improve file coverage");
  }
  const unknownFiles = report.fileCoverage.uncoveredFiles.filter(
    (file) => file.reason.includes("no known linter supports this file type")
  );
  if (unknownFiles.length > 0) {
    steps.push("Review uncovered file types to decide whether new linter support is needed");
  }
  if (report.policy.source === "cloud" && report.policy.totalRules > report.policy.applicableRules) {
    steps.push("Review cloud policy rules that target linters not currently selected in this repo");
  }
  return steps;
}
async function collectRunDecisionReport(options = {}) {
  const rc = loadRC();
  let discoveredFiles;
  let mode;
  if (options.paths && options.paths.length > 0) {
    discoveredFiles = [];
    for (const p of options.paths) {
      discoveredFiles.push(...findFiles(p));
    }
    mode = options.paths.join(", ");
  } else {
    discoveredFiles = getStagedFilePaths();
    mode = "staged files";
  }
  const ignorePatterns = rc.ignore || [];
  const ignoredFiles = getIgnoredFileDecisions(discoveredFiles, ignorePatterns);
  const files = filterIgnoredFiles(discoveredFiles, ignorePatterns);
  const linterSelection = describeLinterSelection(rc);
  const selectedLinters = selectLinters(rc);
  const fileCoverage = describeFileCoverage(files, linterSelection, selectedLinters);
  const conflicts = describeResolvedConflicts(linterSelection);
  let policyRules = [];
  try {
    policyRules = await fetchPolicyRules();
  } catch {
    policyRules = [];
  }
  const policy = summarizePolicyRules(policyRules, selectedLinters);
  return {
    cwd: process.cwd(),
    mode,
    requestedPaths: options.paths ?? [],
    discoveredFileCount: discoveredFiles.length,
    lintableFileCount: files.length,
    ignoredFiles,
    linterSelection,
    fileCoverage,
    policy,
    fixStrategy: {
      autofix: rc.fix?.enabled ?? false,
      strategy: rc.fix?.strategy ?? "parallel"
    },
    conflicts,
    nextSteps: describeNextSteps({
      linterSelection,
      fileCoverage,
      policy
    })
  };
}
function buildMachineSummary(doctorStatus, report, missingSelectedLinters) {
  const actions = [];
  const hasPolicyScopeGap = report.policy.totalRules > report.policy.applicableRules;
  const blockingReasons = [];
  const warningReasons = [];
  if (missingSelectedLinters.length > 0) {
    blockingReasons.push("missing_selected_linters");
    actions.push({
      id: "install_missing_linters",
      label: "Install missing linters",
      command: "lint install missing .",
      reason: `Missing selected linters: ${missingSelectedLinters.join(", ")}`
    });
  }
  if (doctorStatus === "needs_setup") {
    blockingReasons.push("needs_setup");
    actions.push({
      id: "fix_setup",
      label: "Fix repo setup",
      command: "lint setup fix --dry-run",
      reason: "Repository setup is not yet healthy"
    });
  }
  if (report.fileCoverage.uncoveredFiles.length > 0) {
    blockingReasons.push("uncovered_files");
    actions.push({
      id: "explain_run",
      label: "Explain run coverage",
      command: "lint explain run .",
      reason: `${report.fileCoverage.uncoveredFiles.length} file(s) are not covered by a selected linter`
    });
  }
  if (hasPolicyScopeGap) {
    warningReasons.push("policy_scope_gap");
    actions.push({
      id: "review_policy_scope",
      label: "Review policy scope",
      command: "lint explain run --json .",
      reason: "Some cloud policy rules target linters that are not currently selected"
    });
  }
  return {
    status: blockingReasons.length === 0 ? "ready" : "action_required",
    doctor_status: doctorStatus,
    run_mode: report.mode,
    selected_linters: report.linterSelection.filter((entry) => entry.selected).map((entry) => entry.name),
    missing_selected_linters: missingSelectedLinters,
    uncovered_file_count: report.fileCoverage.uncoveredFiles.length,
    ignored_file_count: report.ignoredFiles.length,
    applicable_policy_rule_count: report.policy.applicableRules,
    blocking_reasons: blockingReasons,
    warning_reasons: warningReasons,
    signals: {
      needs_setup: doctorStatus !== "healthy",
      has_missing_selected_linters: missingSelectedLinters.length > 0,
      has_uncovered_files: report.fileCoverage.uncoveredFiles.length > 0,
      has_policy_scope_gap: hasPolicyScopeGap,
      is_actionable: actions.length > 0
    },
    next_steps: report.nextSteps,
    actions,
    primary_action: actions[0] ?? null
  };
}
function getMachineSummaryExitCode(summary) {
  if (summary.status === "action_required") return 1;
  return 0;
}
async function fetchPolicyRules() {
  if (!isLoggedIn()) return [];
  const config = readLintConfig();
  if (!config?.uuid || config.uuid.startsWith("local-")) return [];
  const token = getToken();
  if (!token) return [];
  const result = await fetchPolicy(config.uuid, token);
  return result.data?.policy_rules || [];
}
async function createCommitAttempt2() {
  if (!isLoggedIn()) return null;
  const config = readLintConfig();
  if (!config?.uuid || config.uuid.startsWith("local-")) return null;
  const token = getToken();
  if (!token) return null;
  const sha = getCurrentSha();
  const branch = getCurrentBranch();
  const result = await createCommitAttempt(config.uuid, token, sha, branch);
  return result.data?.id || null;
}
async function sendReport(reports, commitAttemptId) {
  if (!isLoggedIn() || !commitAttemptId) return;
  const config = readLintConfig();
  if (!config?.uuid || config.uuid.startsWith("local-")) return;
  const token = getToken();
  if (!token) return;
  try {
    await postReport(token, {
      commit_attempt_id: commitAttemptId,
      repository_uuid: config.uuid,
      error_count: reports.reduce((s, r) => s + r.error_count, 0),
      warning_count: reports.reduce((s, r) => s + r.warning_count, 0),
      linters: reports.map((r) => ({
        name: r.linter,
        error_count: r.error_count,
        warning_count: r.warning_count
      }))
    });
  } catch {
  }
}
async function runLint(options = {}) {
  const startTime = Date.now();
  const rc = loadRC();
  const autofix = options.fix ?? rc.fix?.enabled ?? false;
  const dryRun = options.dryRun ?? false;
  const verbose = options.verbose ?? false;
  const quiet = options.quiet ?? rc.output?.quiet ?? false;
  const format = options.format ?? rc.output?.format ?? "text";
  const isJson = format === "json";
  const failOnWarnings = options.exitOnWarnings ?? false;
  let files;
  let mode;
  if (options.paths && options.paths.length > 0) {
    files = [];
    for (const p of options.paths) {
      files.push(...findFiles(p));
    }
    mode = options.paths.join(", ");
  } else {
    files = getStagedFilePaths();
    mode = "staged files";
  }
  const ignorePatterns = rc.ignore || [];
  const ignoredFiles = getIgnoredFileDecisions(files, ignorePatterns);
  files = filterIgnoredFiles(files, ignorePatterns);
  const linterSelection = describeLinterSelection(rc);
  const selectedLinters = selectLinters(rc);
  const fileCoverage = describeFileCoverage(files, linterSelection, selectedLinters);
  if (files.length === 0) {
    if (isJson) {
      console.log(
        formatJsonReport([], {
          duration: Date.now() - startTime,
          dryRun,
          fix: autofix,
          cwd: process.cwd(),
          mode,
          fileCount: 0,
          linterNames: [],
          policyRuleCount: 0,
          message: "No files to lint",
          failOnWarnings,
          requestedPaths: options.paths ?? [],
          ignoredFiles,
          linterSelection,
          fileCoverage,
          policySummary: summarizePolicyRules([], selectedLinters)
        })
      );
    } else if (!quiet) {
      console.log(chalk11.green("No files to lint."));
    }
    return;
  }
  if (!quiet && !isJson) {
    const action = dryRun ? "Checking" : autofix ? "Fixing" : "Linting";
    console.log(chalk11.cyan(`
Lint \u2014 ${action} ${files.length} file(s) (${mode})...
`));
  }
  let policyRules = [];
  try {
    policyRules = await fetchPolicyRules();
    if (verbose && !isJson && policyRules.length > 0) {
      console.log(chalk11.gray(`  Loaded ${policyRules.length} policy rule(s) from API`));
    }
  } catch {
    if (verbose && !isJson) console.log(chalk11.gray("  API offline \u2014 using linter defaults"));
  }
  const commitAttemptId = await createCommitAttempt2().catch(() => null);
  const linters = selectedLinters;
  if (linters.length === 0) {
    if (isJson) {
      console.log(
        formatJsonReport([], {
          duration: Date.now() - startTime,
          dryRun,
          fix: autofix,
          cwd: process.cwd(),
          mode,
          fileCount: files.length,
          linterNames: [],
          policyRuleCount: policyRules.length,
          message: "No linters available",
          failOnWarnings,
          requestedPaths: options.paths ?? [],
          ignoredFiles,
          linterSelection,
          fileCoverage,
          policySummary: summarizePolicyRules(policyRules, linters)
        })
      );
    } else if (!quiet) {
      console.log(
        chalk11.yellow("No linters available. Run 'lint init' to set up, or install one manually.")
      );
    }
    return;
  }
  if (verbose && !isJson) {
    console.log(chalk11.gray(`  Linters: ${linters.map((l) => l.name).join(", ")}
`));
  }
  const results = [];
  let hasErrors = false;
  const jsLinters = linters.filter(
    (l) => ["biome", "oxlint", "eslint", "prettier"].includes(l.name)
  );
  const otherLinters = linters.filter(
    (l) => !["biome", "oxlint", "eslint", "prettier"].includes(l.name)
  );
  const runLinter = (linter) => {
    const relevantFiles = linter.selectFiles(files);
    if (relevantFiles.length === 0) return null;
    if (!quiet && !isJson) {
      const spinner = createSpinner(`${linter.name} (${relevantFiles.length} files)`).start();
      const result = linter.execute(files, policyRules, dryRun ? false : autofix);
      if (!result) {
        spinner.success({ text: `${linter.name} \u2014 skipped` });
        return null;
      }
      if (result.report.error_count > 0) {
        spinner.error({
          text: `${linter.name} \u2014 ${result.report.error_count} error(s), ${result.report.warning_count} warning(s)`
        });
      } else if (result.report.warning_count > 0) {
        spinner.warn({ text: `${linter.name} \u2014 ${result.report.warning_count} warning(s)` });
      } else {
        spinner.success({ text: `${linter.name} \u2014 clean` });
      }
      return result;
    }
    return linter.execute(files, policyRules, dryRun ? false : autofix);
  };
  const fixStrategy = rc.fix?.strategy ?? "parallel";
  if (fixStrategy === "formatter-first" && autofix) {
    const formatters = jsLinters.filter((l) => ["prettier", "biome"].includes(l.name));
    const nonFormatters = jsLinters.filter((l) => !["prettier", "biome"].includes(l.name));
    for (const linter of [...formatters, ...nonFormatters]) {
      const result = runLinter(linter);
      if (result) {
        results.push(result);
        if (result.report.error_count > 0) hasErrors = true;
      }
    }
  } else {
    for (const linter of jsLinters) {
      const result = runLinter(linter);
      if (result) {
        results.push(result);
        if (result.report.error_count > 0) hasErrors = true;
      }
    }
  }
  const otherResults = await Promise.all(otherLinters.map(async (linter) => runLinter(linter)));
  for (const result of otherResults) {
    if (result) {
      results.push(result);
      if (result.report.error_count > 0) hasErrors = true;
    }
  }
  const reports = results.map((r) => r.report);
  const totalErrors = reports.reduce((s, r) => s + r.error_count, 0);
  const totalWarnings = reports.reduce((s, r) => s + r.warning_count, 0);
  const totalFixable = reports.reduce(
    (s, r) => s + r.fixable_error_count + r.fixable_warning_count,
    0
  );
  const duration = Date.now() - startTime;
  if (isJson) {
    console.log(
      formatJsonReport(reports, {
        duration,
        dryRun,
        fix: autofix,
        cwd: process.cwd(),
        mode,
        fileCount: files.length,
        linterNames: linters.map((linter) => linter.name),
        policyRuleCount: policyRules.length,
        failOnWarnings,
        requestedPaths: options.paths ?? [],
        ignoredFiles,
        linterSelection,
        fileCoverage,
        policySummary: summarizePolicyRules(policyRules, linters)
      })
    );
  } else {
    if (!quiet && reports.some((r) => r.error_count > 0 || r.warning_count > 0)) {
      printReport(reports, options.truncate);
      if (verbose) printSummaryTable(reports);
    }
    console.log("");
    if (dryRun && totalFixable > 0) {
      console.log(
        chalk11.cyan(`  ${totalFixable} issue(s) would be fixed. Run without --dry-run to apply.`)
      );
    }
    if (totalErrors > 0) {
      console.log(chalk11.red(`\u2717 ${totalErrors} error(s), ${totalWarnings} warning(s)`));
      if (totalFixable > 0 && !autofix) {
        console.log(chalk11.yellow(`  ${totalFixable} auto-fixable. Run 'lint --fix' to fix.`));
      }
    } else if (totalWarnings > 0) {
      console.log(chalk11.yellow(`\u26A0 ${totalWarnings} warning(s)`));
    } else {
      console.log(chalk11.green("\u2713 All files passed."));
    }
    if (options.time) {
      console.log(chalk11.gray(`  Done in ${formatDuration(duration)}`));
    }
  }
  await sendReport(reports, commitAttemptId);
  if (!options.keep) cleanTmpDir();
  if (hasErrors) process.exit(1);
  if (totalWarnings > 0 && failOnWarnings) process.exit(2);
}
async function explainRun(options = {}) {
  const report = await collectRunDecisionReport(options);
  if (options.json) {
    console.log(formatRunDecisionJson(report));
    return;
  }
  for (const line of formatRunDecisionReport(report)) {
    console.log(line);
  }
}
async function machineSummary(options) {
  const report = await collectRunDecisionReport(options);
  const summary = buildMachineSummary(options.doctorStatus, report, options.missingSelectedLinters);
  console.log(formatMachineSummaryJson(summary));
  if (options.strict) {
    const exitCode = getMachineSummaryExitCode(summary);
    if (exitCode !== 0) process.exit(exitCode);
  }
}
async function preCommit(options = {}) {
  return runLint({ ...options, time: options.time ?? true });
}
async function prepareCommitMsg() {
}
async function postCommitHook() {
}
async function formatProjectFiles(extension) {
  const prettier = new PrettierLinter();
  if (!prettier.isInstalled()) {
    console.log(chalk11.red("Prettier is not installed. Run: npm install -g prettier"));
    return;
  }
  const spinner = createSpinner(`Formatting all .${extension} files...`).start();
  try {
    execFile("prettier", ["--write", `**/*.${extension}`], { silent: true });
    spinner.success({ text: `All .${extension} files formatted.` });
  } catch (error) {
    const err = error;
    spinner.error({ text: `Prettier failed: ${err.stderr || err.message}` });
  }
}

// src/index.ts
var program = new Command();
program.name("lint").description("Lint \u2014 The universal linter with AI-powered code review.").version(VERSION, "-v, --version");
function addHiddenCommand(command) {
  program.addCommand(command, { hidden: true });
}
function addLegacyAlias(name, description, configure) {
  addHiddenCommand(configure(new Command(name).description(description)));
}
function withJsonOption(command, description) {
  return command.option("--json", description);
}
function withBootstrapOptions(command) {
  return command.option("--dry-run", "Preview the bootstrap plan without writing files").option("--json", "Output the bootstrap plan as JSON").option("--install-missing", "Install suggested missing linters").option("--install-hooks", "Install managed git hooks");
}
function withSetupFixOptions(command) {
  return command.option("--dry-run", "Preview the setup changes without writing files").option("--json", "Output the setup plan as JSON").option("--no-install-missing", "Skip installing missing suggested linters").option("--no-install-hooks", "Skip installing managed git hooks");
}
program.command("check [paths...]", { isDefault: true }).description("Lint files (default: staged files, or specify paths)").option("--fix", "Auto-fix issues where supported").option("--dry-run", "Show what would be fixed without applying").option("-f, --format <format>", "Output format: text, json", "text").option("-q, --quiet", "Minimal output (just summary)").option("--verbose", "Detailed output").option("-T, --truncate", "Truncate to first 10 offenses per file").option("-t, --time", "Show execution time").option("--exit-on-warnings", "Exit with code 2 on warnings").action((paths, options) => {
  runLint({
    paths: paths.length > 0 ? paths : void 0,
    fix: options.fix,
    dryRun: options.dryRun,
    format: options.format,
    quiet: options.quiet,
    verbose: options.verbose,
    truncate: options.truncate,
    time: options.time,
    exitOnWarnings: options.exitOnWarnings
  });
});
program.command("ci [paths...]").description("Run a repo-local quality gate for CI or control planes").option("--fix", "Auto-fix issues where supported").option("--dry-run", "Show what would be fixed without applying").option("-f, --format <format>", "Output format: json, text", "json").option("-q, --quiet", "Minimal output (just summary)").option("--verbose", "Detailed output").option("-T, --truncate", "Truncate to first 10 offenses per file").option("-t, --time", "Show execution time").option("--allow-warnings", "Exit 0 when only warnings are found").action((paths, options) => {
  runLint({
    paths: paths.length > 0 ? paths : ["."],
    fix: options.fix,
    dryRun: options.dryRun,
    format: options.format,
    quiet: options.quiet,
    verbose: options.verbose,
    truncate: options.truncate,
    time: options.time,
    exitOnWarnings: !options.allowWarnings
  });
});
function installManagedHooks() {
  console.log("Installing git hooks...");
  const rc = loadRC();
  installHooks({ timeout: rc.hooks?.timeout, skipEnv: rc.hooks?.skip_env });
}
function runPreCommitHook(options) {
  preCommit({
    fix: options.fix,
    time: options.time ?? true,
    truncate: options.truncate,
    quiet: options.quiet
  });
}
function runPrepareCommitMsgHook() {
  prepareCommitMsg();
}
function runPostCommitCommand() {
  postCommitHook();
}
function runMachineSummary(paths, options) {
  const doctor = collectDoctorReport();
  machineSummary({
    paths: paths.length > 0 ? paths : ["."],
    doctorStatus: doctor.status,
    missingSelectedLinters: doctor.summary.missingSelectedLinters,
    strict: options.strict
  });
}
function runSetupFix(options) {
  fixSetup({
    dryRun: options.dryRun,
    json: options.json,
    installMissing: options.installMissing,
    installHooks: options.installHooks
  });
}
function runSetupBootstrap(options) {
  bootstrapProject({
    dryRun: options.dryRun,
    json: options.json,
    installMissing: options.installMissing,
    installHooks: options.installHooks
  });
}
function runDoctor(options) {
  const report = collectDoctorReport();
  if (options.json) {
    console.log(JSON.stringify(report, null, 2));
    return;
  }
  console.log(chalk12.cyan.bold(""));
  for (const line of formatDoctorReport(report)) {
    console.log(line);
  }
}
function runExplainRun(paths, options) {
  explainRun({
    paths: paths.length > 0 ? paths : ["."],
    json: options.json
  });
}
function runRecommendedConfig(options) {
  const project = detectProject(findGitRoot() || process.cwd());
  const suggested = buildSuggestedLinterPlan(project).map((entry) => entry.name);
  const existing = loadRC();
  const recommended = buildRecommendedRC(existing, suggested);
  if (options.write) {
    writeRC(recommended);
    console.log(chalk12.green("Wrote recommended .lintrc.yaml"));
    return;
  }
  if (options.json) {
    console.log(
      JSON.stringify(
        {
          schema_version: LINT_JSON_SCHEMA_VERSION,
          kind: "lint_recommended_config",
          suggested_linters: suggested,
          recommended
        },
        null,
        2
      )
    );
    return;
  }
  console.log(chalk12.cyan.bold("\n  Recommended .lintrc.yaml\n"));
  console.log(formatRC(recommended));
}
function runInstallMissing(_paths, options) {
  const root = findGitRoot() || process.cwd();
  const project = detectProject(root);
  const plan = buildSuggestedLinterPlan(project);
  const missing = plan.filter((entry) => !entry.installed);
  if (missing.length === 0) {
    console.log(chalk12.green("No suggested linters are missing."));
    return;
  }
  console.log(chalk12.bold("\n  Missing Suggested Linters\n"));
  for (const entry of missing) {
    console.log(`  - ${entry.name}`);
    if (entry.reasons.length > 0) {
      console.log(chalk12.gray(`    ${entry.reasons.join(", ")}`));
    }
  }
  console.log("");
  if (options.dryRun) {
    return;
  }
  for (const entry of missing) {
    const linter = LINTER_MAP.get(entry.name);
    if (!linter) continue;
    linter.install();
  }
}
function printHooksStatus(options) {
  const gitRoot = findGitRoot();
  if (!gitRoot) {
    console.log(chalk12.red("Not inside a git repository."));
    return;
  }
  const hooks = inspectManagedHooks(gitRoot);
  if (options.json) {
    console.log(JSON.stringify(hooks, null, 2));
    return;
  }
  console.log(chalk12.bold("\n  Lint Hooks\n"));
  for (const hook of hooks) {
    const status = !hook.exists ? chalk12.gray("missing") : hook.managed ? chalk12.green("managed") : chalk12.yellow("unmanaged");
    console.log(`  ${hook.name}: ${status}`);
    if (hook.exists) {
      console.log(chalk12.gray(`    ${hook.hookPath}`));
    }
  }
  console.log("");
}
async function runLogin() {
  if (isLoggedIn()) {
    console.log(`Already logged in as ${chalk12.green(getUsername())}.`);
    return;
  }
  const username = await input2({ message: "Username or email:" });
  const pw = await password({ message: "Password:", mask: "*" });
  await login2(username, pw);
}
async function runLogout() {
  if (!isLoggedIn()) {
    console.log("Not logged in.");
    return;
  }
  const yes = await confirm2({ message: "Are you sure you want to log out?" });
  if (yes) logout();
}
async function runSignup() {
  if (isLoggedIn()) {
    console.log(`Already logged in as ${chalk12.green(getUsername())}.`);
    return;
  }
  const username = await input2({ message: "Username:" });
  const email = await input2({ message: "Email:" });
  const pw = await password({ message: "Password:", mask: "*" });
  await signup2(username, email, pw);
}
function runAuthStatus() {
  printStatus();
}
function runFormatWrite(extension) {
  formatProjectFiles(extension);
}
addHiddenCommand(
  new Command("pre-commit").description("Pre-commit hook: lint staged files").option("--fix", "Auto-fix issues").option("-t, --time", "Show execution time").option("-T, --truncate", "Truncate output").option("-q, --quiet", "Minimal output").action((options) => runPreCommitHook(options))
);
addHiddenCommand(
  new Command("prepare-commit-msg").description("Prepare-commit-msg hook").action(() => runPrepareCommitMsgHook())
);
addHiddenCommand(
  new Command("post-commit").description("Post-commit hook").action(() => runPostCommitCommand())
);
var hooksCommand = program.command("hooks").description("Manage Lint git hooks");
hooksCommand.command("install").description("Install git hooks").action(() => installManagedHooks());
hooksCommand.command("uninstall").description("Remove Lint git hooks").action(() => uninstallHooks());
hooksCommand.command("status").description("Inspect Lint git hooks").option("--json", "Output hook status as JSON").action((options) => printHooksStatus(options));
program.command("init").description("Initialize Lint with smart project detection").action(() => init());
withBootstrapOptions(
  program.command("bootstrap").description("Bootstrap repo-local Lint config without interactive prompts")
).action(
  (options) => runSetupBootstrap(options)
);
program.command("doctor").description("Diagnose Lint setup and linter health").option("--json", "Output doctor status as JSON").action((options) => runDoctor(options));
var setupCommand = program.command("setup").description("Apply or repair repo-local Lint setup");
var setupFixCommand = withSetupFixOptions(
  setupCommand.command("fix").description("Apply recommended repo-local Lint setup in one pass")
);
setupFixCommand.action(
  (options) => runSetupFix(options)
);
var configCommand = program.command("config").description("Manage Lint configuration helpers");
configCommand.command("recommend").description("Show or write a recommended .lintrc.yaml for this project").option("--json", "Output the recommendation as JSON").option("--write", "Write the recommended config to .lintrc.yaml").action((options) => runRecommendedConfig(options));
var installCommand = program.command("install").description("Install Lint-related tooling");
installCommand.command("missing [paths...]").description("Install missing linters suggested for this project").option("--dry-run", "Show the missing linters without installing them").action((paths, options) => runInstallMissing(paths, options));
var machineCommand = program.command("machine").description("Machine-readable Lint output helpers");
machineCommand.command("summary [paths...]").description("Output a compact machine-readable repo summary for automation consumers").option("--strict", "Exit 1 when the repo still needs setup or has uncovered files").action((paths, options) => runMachineSummary(paths, options));
var explainCommand = program.command("explain").description("Explain Lint decisions and behavior");
explainCommand.command("run [paths...]").description("Explain file, linter, and policy decisions for a run without linting").option("--json", "Output explanation as JSON").action((paths, options) => runExplainRun(paths, options));
var authCommand = program.command("auth").description("Manage Lint authentication");
authCommand.command("status").description("Show current login status").action(() => runAuthStatus());
authCommand.command("login").description("Sign in to Lint").action(async () => runLogin());
authCommand.command("logout").description("Sign out from Lint").action(async () => runLogout());
authCommand.command("signup").description("Create a Lint account").action(async () => runSignup());
authCommand.addCommand(
  new Command("whoami").description("Legacy alias for 'lint auth status'").action(() => runAuthStatus()),
  { hidden: true }
);
var formatCommand = program.command("format").description("Formatting helpers");
formatCommand.command("write <extension>").description("Run Prettier on all files with the given extension").action((extension) => runFormatWrite(extension));
addLegacyAlias(
  "explain-run [paths...]",
  "Legacy alias for 'lint explain run'",
  (command) => withJsonOption(command, "Output explanation as JSON").action(
    (paths, options) => runExplainRun(paths, options)
  )
);
setupCommand.addCommand(
  new Command("init").description("Legacy alias for 'lint init'").action(() => init()),
  { hidden: true }
);
setupCommand.addCommand(
  withBootstrapOptions(
    new Command("bootstrap").description("Legacy alias for 'lint bootstrap'")
  ).action(
    (options) => runSetupBootstrap(options)
  ),
  { hidden: true }
);
setupCommand.addCommand(
  withJsonOption(
    new Command("doctor").description("Legacy alias for 'lint doctor'"),
    "Output doctor status as JSON"
  ).action((options) => runDoctor(options)),
  { hidden: true }
);
addLegacyAlias(
  "install:hooks",
  "Legacy alias for 'lint hooks install'",
  (command) => command.action(() => installManagedHooks())
);
addLegacyAlias(
  "machine:summary [paths...]",
  "Legacy alias for 'lint machine summary'",
  (command) => command.option("--strict", "Exit 1 when the repo still needs setup or has uncovered files").action((paths, options) => runMachineSummary(paths, options))
);
addLegacyAlias(
  "setup:fix",
  "Legacy alias for 'lint setup fix'",
  (command) => withSetupFixOptions(command).action(
    (options) => runSetupFix(options)
  )
);
addLegacyAlias(
  "config:recommend",
  "Legacy alias for 'lint config recommend'",
  (command) => withJsonOption(command, "Output the recommendation as JSON").option("--write", "Write the recommended config to .lintrc.yaml").action((options) => runRecommendedConfig(options))
);
addLegacyAlias(
  "install:missing [paths...]",
  "Legacy alias for 'lint install missing'",
  (command) => command.option("--dry-run", "Show the missing linters without installing them").action((paths, options) => runInstallMissing(paths, options))
);
addLegacyAlias(
  "uninstall:hooks",
  "Legacy alias for 'lint hooks uninstall'",
  (command) => command.action(() => uninstallHooks())
);
addLegacyAlias(
  "hooks:status",
  "Legacy alias for 'lint hooks status'",
  (command) => withJsonOption(command, "Output hook status as JSON").action(
    (options) => printHooksStatus(options)
  )
);
addLegacyAlias(
  "login",
  "Legacy alias for 'lint auth login'",
  (command) => command.action(async () => runLogin())
);
addLegacyAlias(
  "logout",
  "Legacy alias for 'lint auth logout'",
  (command) => command.action(async () => runLogout())
);
addLegacyAlias(
  "signup",
  "Legacy alias for 'lint auth signup'",
  (command) => command.action(async () => runSignup())
);
addLegacyAlias(
  "whoami",
  "Legacy alias for 'lint auth whoami'",
  (command) => command.action(() => runAuthStatus())
);
addLegacyAlias(
  "prettify <extension>",
  "Legacy alias for 'lint format write <extension>'",
  (command) => command.action((extension) => runFormatWrite(extension))
);
var ai = program.command("ai").description("AI-powered code analysis (powered by Claude)");
ai.command("review").description("AI code review of staged changes").action(() => reviewStagedChanges());
ai.command("fix").description("AI-powered auto-fix suggestions").action(() => fixStagedChanges());
ai.command("commit").description("Generate a commit message from staged changes").action(() => printCommitSuggestion());
ai.command("explain").description("Explain linting errors in plain language").action(async () => {
  console.log(chalk12.cyan("Running linters to collect errors...\n"));
  const files = getStagedFilePaths();
  if (files.length === 0) {
    console.log(chalk12.yellow("No staged files."));
    return;
  }
  const reports = [];
  for (const linter of ALL_LINTERS2) {
    if (!linter.isInstalled()) continue;
    if (linter.selectFiles(files).length === 0) continue;
    const result = linter.execute(files, [], false);
    if (result?.report) reports.push(result.report);
  }
  await explainErrors(reports);
});
ai.command("setup").description("Configure your Anthropic API key").action(async () => {
  const apiKey = await input2({
    message: "Enter your Anthropic API key (from https://console.anthropic.com/):"
  });
  if (!apiKey.startsWith("sk-")) {
    console.log(chalk12.red("Invalid API key format. Keys start with 'sk-'."));
    return;
  }
  saveApiKey(apiKey);
  console.log(chalk12.green("API key saved. AI features are now enabled."));
});
var entryArg = process.argv[1];
if (entryArg) {
  let entryHref;
  try {
    entryHref = pathToFileURL(realpathSync(entryArg)).href;
  } catch {
    entryHref = pathToFileURL(entryArg).href;
  }
  if (import.meta.url === entryHref) {
    program.parse();
  }
}
export {
  program
};
